<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Parallel and Memory-Efficient Ecological Diversity Metrics • ecodive</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="katex-auto.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/Noto_Sans-0.4.10/font.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Parallel and Memory-Efficient Ecological Diversity Metrics">
<meta name="description" content="Computes alpha and beta diversity metrics using concurrent C threads. Metrics include UniFrac, Faiths phylogenetic diversity, Bray-Curtis dissimilarity, Shannon diversity index, and many others. Also parses newick trees into phylo' objects and rarefies feature tables.">
<meta property="og:description" content="Computes alpha and beta diversity metrics using concurrent C threads. Metrics include UniFrac, Faiths phylogenetic diversity, Bray-Curtis dissimilarity, Shannon diversity index, and many others. Also parses newick trees into phylo' objects and rarefies feature tables.">
<meta property="og:image" content="https://cmmr.github.io/ecodive/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0.9001</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/ecodive.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="articles/unifrac.html">UniFrac Calculations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="ecodive-">ecodive <a class="anchor" aria-label="anchor" href="#ecodive-"></a>
</h1>
</div>
<!-- badges: start -->

<p><code>ecodive</code> is an R package for calculating ecological diversity metrics in a parallelized and memory-efficient manner. It is designed to handle large datasets, such as those common in microbiome research, with significant speed and memory improvements over existing tools.</p>
<div class="section level2">
<h2 id="why-ecodive">Why ecodive?<a class="anchor" aria-label="anchor" href="#why-ecodive"></a>
</h2>
<p>Analyzing ecological diversity is often a computational bottleneck, especially with large datasets. <code>ecodive</code> addresses this by providing:</p>
<ul>
<li><p><strong>High Performance:</strong> <code>ecodive</code> is written in C and parallelized using pthreads, making it dramatically faster than other R packages. <a href="https://cmmr.github.io/ecodive/articles/benchmark.html">Benchmarks</a> show it can be up to <strong>43,000x faster</strong> and use up to <strong>33,000x less memory</strong>.</p></li>
<li><p><strong>Zero Dependencies:</strong> The package has no external R dependencies, making it lightweight, stable, and easy to install. This also makes it an ideal and secure backend for other R packages.</p></li>
<li><p><strong>Comprehensive Metrics:</strong> It implements a wide range of common alpha and beta diversity metrics, including classic and phylogenetic-aware methods like <strong>Faith’s PD</strong> and the <strong>complete UniFrac family</strong>.</p></li>
<li><p><strong>Ease of Use:</strong> The API is simple and integrates seamlessly with popular bioinformatics packages like <a href="http://joey711.github.io/phyloseq/" class="external-link"><code>phyloseq</code></a> and <a href="https://cmmr.github.io/rbiom/" class="external-link"><code>rbiom</code></a>.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest stable version can be installed from CRAN.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'ecodive'</span><span class="op">)</span></span></code></pre></div>
<p>The development version is available on GitHub. Please note that this method requires a compiler - see <a href="http://www.rstudio.com/ide/docs/packages/prerequisites" class="external-link uri">http://www.rstudio.com/ide/docs/packages/prerequisites</a> if the installation does not succeed on the first try.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'pak'</span><span class="op">)</span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pak.html" class="external-link">pak</a></span><span class="op">(</span><span class="st">'cmmr/ecodive'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p><code>ecodive</code> functions are straightforward to use. Here are a few examples.</p>
<div class="section level3">
<h3 id="with-phyloseq-or-rbiom-objects">With <code>phyloseq</code> or <code>rbiom</code> objects<a class="anchor" aria-label="anchor" href="#with-phyloseq-or-rbiom-objects"></a>
</h3>
<p>The easiest way to use <code>ecodive</code> is with a <code>phyloseq</code> or <code>rbiom</code> object. These objects conveniently bundle the count data and phylogenetic tree.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/ecodive/">ecodive</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">esophagus</span>, package <span class="op">=</span> <span class="st">'phyloseq'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hmp50</span>,     package <span class="op">=</span> <span class="st">'rbiom'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate weighted UniFrac distance</span></span>
<span><span class="va">w_unifrac</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/weighted_unifrac.html">weighted_unifrac</a></span><span class="op">(</span><span class="va">esophagus</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">w_unifrac</span><span class="op">)</span></span>
<span><span class="co">#&gt;           B         C</span></span>
<span><span class="co">#&gt; C 0.1050480          </span></span>
<span><span class="co">#&gt; D 0.1401124 0.1422409</span></span>
<span></span>
<span><span class="co"># Calculate Faith's Phylogenetic Diversity</span></span>
<span><span class="va">faith_pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/faith.html">faith</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">faith_pd</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;   HMP01   HMP02   HMP03   HMP04 </span></span>
<span><span class="co">#&gt; 6.22296 8.59432 8.93375 9.86597 </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="with-basic-r-objects">With basic R objects<a class="anchor" aria-label="anchor" href="#with-basic-r-objects"></a>
</h3>
<p>You can also provide the count data and phylogenetic tree as separate objects.</p>
<p>The <code>ex_counts</code> and <code>ex_tree</code> objects are included with <code>ecodive</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example Data ----------------------</span></span>
<span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span>
<span><span class="va">counts</span></span>
<span><span class="co">#&gt;                   Saliva Gums Nose Stool</span></span>
<span><span class="co">#&gt; Streptococcus        162  309    6     1</span></span>
<span><span class="co">#&gt; Bacteroides            2    2    0   341</span></span>
<span><span class="co">#&gt; Corynebacterium        0    0  171     1</span></span>
<span><span class="co">#&gt; Haemophilus          180   34    0     1</span></span>
<span><span class="co">#&gt; Propionibacterium      1    0   82     0</span></span>
<span><span class="co">#&gt; Staphylococcus         0    0   86     1</span></span>
<span></span>
<span></span>
<span><span class="co">## Alpha Diversity -------------------</span></span>
<span></span>
<span><span class="fu"><a href="reference/shannon.html">shannon</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Saliva       Gums       Nose      Stool </span></span>
<span><span class="co">#&gt; 0.74119910 0.35692121 1.10615349 0.07927797 </span></span>
<span></span>
<span><span class="fu"><a href="reference/faith.html">faith</a></span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Saliva   Gums   Nose  Stool </span></span>
<span><span class="co">#&gt;    180    155    101    202 </span></span>
<span></span>
<span></span>
<span><span class="co">## Beta Diversity --------------------</span></span>
<span></span>
<span><span class="fu"><a href="reference/bray_curtis.html">bray_curtis</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Saliva      Gums      Nose</span></span>
<span><span class="co">#&gt; Gums  0.4260870                    </span></span>
<span><span class="co">#&gt; Nose  0.9797101 0.9826087          </span></span>
<span><span class="co">#&gt; Stool 0.9884058 0.9884058 0.9913043</span></span>
<span></span>
<span><span class="fu"><a href="reference/weighted_unifrac.html">weighted_unifrac</a></span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Saliva      Gums      Nose</span></span>
<span><span class="co">#&gt; Gums   36.97681                    </span></span>
<span><span class="co">#&gt; Nose   67.23768  55.97101          </span></span>
<span><span class="co">#&gt; Stool 109.77971 109.44058 110.00870</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>The online manual for ecodive is available at <a href="https://cmmr.github.io/ecodive/" class="uri">https://cmmr.github.io/ecodive/</a>. It includes a getting started guide, articles on alpha/beta diversity, and reference pages for each function.</p>
</div>
<div class="section level2">
<h2 id="automated-tests">Automated tests<a class="anchor" aria-label="anchor" href="#automated-tests"></a>
</h2>
<p>The following commands will check if ecodive passes the bundled testing suite.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'testthat'</span><span class="op">)</span></span>
<span><span class="fu">testthat</span><span class="fu">::</span><span class="fu"><a href="https://testthat.r-lib.org/reference/test_package.html" class="external-link">test_check</a></span><span class="op">(</span><span class="st">'ecodive'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="community-guidelines">Community guidelines<a class="anchor" aria-label="anchor" href="#community-guidelines"></a>
</h2>
<div class="section level3">
<h3 id="support">Support<a class="anchor" aria-label="anchor" href="#support"></a>
</h3>
<p>Bug reports, feature requests, and general questions can be submitted at <a href="https://github.com/cmmr/ecodive/issues" class="external-link uri">https://github.com/cmmr/ecodive/issues</a>.</p>
</div>
<div class="section level3">
<h3 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h3>
<p>Pull requests are welcome. Please ensure contributed code is covered by tests and documentation (add additional tests and documentation as needed) and passes all automated tests.</p>
<p>New functions must leverage C and pthreads to minimize memory and CPU time.</p>
<p>Please note that the ecodive project is released with a <a href="https://cmmr.github.io/ecodive/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cmmr/ecodive/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/cmmr/ecodive/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ecodive</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Daniel P. Smith <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2479-2044" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Alkek Center for Metagenomics and Microbiome Research <br><small class="roles"> Copyright holder, funder </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cmmr/ecodive/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/cmmr/ecodive/actions/workflows/R-CMD-check.yaml/badge.svg" alt="dev"></a></li>
<li><a href="https://app.codecov.io/gh/cmmr/ecodive" class="external-link"><img src="https://codecov.io/gh/cmmr/ecodive/graph/badge.svg" alt="covr"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
