<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Beta Diversity Metrics — bdiv_functions • ecodive</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Beta Diversity Metrics — bdiv_functions"><meta name="description" content="Beta Diversity Metrics"><meta property="og:description" content="Beta Diversity Metrics"><meta property="og:image" content="https://cmmr.github.io/ecodive/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Beta Diversity Metrics</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/v2.1.0/R/beta_div.r" class="external-link"><code>R/beta_div.r</code></a></small>
      <div class="d-none name"><code>bdiv_functions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Beta Diversity Metrics</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">aitchison</span><span class="op">(</span><span class="va">counts</span>, pseudocount <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">bhattacharyya</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">bray</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">canberra</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">chebyshev</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">chord</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">clark</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">divergence</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">euclidean</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gower</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hellinger</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">horn</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">jensen</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">jsd</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lorentzian</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">manhattan</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">matusita</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">minkowski</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, power <span class="op">=</span> <span class="fl">1.5</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">morisita</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">motyka</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">psym_chisq</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">soergel</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">squared_chisq</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">squared_chord</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">squared_euclidean</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">topsoe</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">wave_hedges</span><span class="op">(</span><span class="va">counts</span>, norm <span class="op">=</span> <span class="st">"percent"</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hamming</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">jaccard</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ochiai</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sorensen</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">unweighted_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">weighted_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">normalized_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">generalized_unifrac</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  tree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  pairs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">variance_adjusted_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>A numeric matrix of count data where each column is a
feature, and each row is a sample. Any object coercible with
<code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code> can be given here, as well as <code>phyloseq</code>, <code>rbiom</code>,
<code>SummarizedExperiment</code>, and <code>TreeSummarizedExperiment</code> objects.</p></dd>


<dt id="arg-pseudocount">pseudocount<a class="anchor" aria-label="anchor" href="#arg-pseudocount"></a></dt>
<dd><p>The value to add to all counts in <code>counts</code> to prevent
taking <code>log(0)</code> for unobserved features. The default, <code>NULL</code>, selects
the smallest non-zero value in <code>counts</code>.</p></dd>


<dt id="arg-pairs">pairs<a class="anchor" aria-label="anchor" href="#arg-pairs"></a></dt>
<dd><p>Which combinations of samples should distances be
calculated for? The default value (<code>NULL</code>) calculates all-vs-all.
Provide a numeric or logical vector specifying positions in the
distance matrix to calculate. See examples.</p></dd>


<dt id="arg-cpus">cpus<a class="anchor" aria-label="anchor" href="#arg-cpus"></a></dt>
<dd><p>How many parallel processing threads should be used. The
default, <code><a href="n_cpus.html">n_cpus()</a></code>, will use all logical CPU cores.</p></dd>


<dt id="arg-norm">norm<a class="anchor" aria-label="anchor" href="#arg-norm"></a></dt>
<dd><p>Normalize the incoming counts. Options are:</p><dl><dt><code>norm = "percent"</code> - </dt>
<dd><p>Relative abundance (sample abundances sum to 1).</p></dd>

<dt><code>norm = "binary"</code>  - </dt>
<dd><p>Unweighted presence/absence (each count is either 0 or 1).</p></dd>

<dt><code>norm = "clr"</code>     - </dt>
<dd><p>Centered log ratio.</p></dd>

<dt><code>norm = "none"</code>    - </dt>
<dd><p>No transformation.</p></dd>


</dl><p>Default: <code>'percent'</code>, which is the expected input for these formulas.</p></dd>


<dt id="arg-power">power<a class="anchor" aria-label="anchor" href="#arg-power"></a></dt>
<dd><p>Scaling factor for the magnitude of differences between
communities (\(p\)). Default: <code>1.5</code></p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>A <code>phylo</code>-class object representing the phylogenetic tree for
the OTUs in <code>counts</code>. The OTU identifiers given by <code>colnames(counts)</code>
must be present in <code>tree</code>. Can be omitted if a tree is embedded with
the <code>counts</code> object or as <code>attr(counts, 'tree')</code>.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>How much weight to give to relative abundances; a value
between 0 and 1, inclusive. Setting <code>alpha=1</code> is equivalent to
<code>normalized_unifrac()</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>dist</code> object.</p>
    </div>
    <div class="section level2">
    <h2 id="formulas">Formulas<a class="anchor" aria-label="anchor" href="#formulas"></a></h2>



<p>Given:</p><ul><li><p>\(n\) : The number of features.</p></li>
<li><p>\(X_i\), \(Y_i\) : Absolute counts for the \(i\)-th feature in samples \(X\) and \(Y\).</p></li>
<li><p>\(X_T\), \(Y_T\) : Total counts in each sample. \(X_T = \sum_{i=1}^{n} X_i\)</p></li>
<li><p>\(P_i\), \(Q_i\) : Proportional abundances of \(X_i\) and \(Y_i\). \(P_i = X_i / X_T\)</p></li>
<li><p>\(X_L\), \(Y_L\) : Mean log of abundances. \(X_L = \frac{1}{n}\sum_{i=1}^{n} \ln{X_i}\)</p></li>
<li><p>\(R_i\) : The range of the \(i\)-th feature across all samples (max - min).</p></li>
</ul><table class="table table"><tr><td></td><td></td></tr><tr><td><strong>Aitchison distance</strong>                            <br><code>aitchison()</code></td><td>\(\sqrt{\sum_{i=1}^{n} [(\ln{X_i} - X_L) - (\ln{Y_i} - Y_L)]^2}\)</td></tr><tr><td><strong>Bhattacharyya distance</strong>                        <br><code>bhattacharyya()</code></td><td>\(-\ln{\sum_{i=1}^{n}\sqrt{P_{i}Q_{i}}}\)</td></tr><tr><td><strong>Bray-Curtis dissimilarity</strong>                     <br><code>bray()</code></td><td>\(\displaystyle \frac{\sum_{i=1}^{n} |P_i - Q_i|}{\sum_{i=1}^{n} (P_i + Q_i)}\)</td></tr><tr><td><strong>Canberra distance</strong>                             <br><code>canberra()</code></td><td>\(\displaystyle \sum_{i=1}^{n} \frac{|P_i - Q_i|}{P_i + Q_i}\)</td></tr><tr><td><strong>Chebyshev distance</strong>                            <br><code>chebyshev()</code></td><td>\(\max(|P_i - Q_i|)\)</td></tr><tr><td><strong>Chord distance</strong>                                <br><code>chord()</code></td><td>\(\displaystyle \sqrt{\sum_{i=1}^{n} \left(\frac{X_i}{\sqrt{\sum_{j=1}^{n} X_j^2}} - \frac{Y_i}{\sqrt{\sum_{j=1}^{n} Y_j^2}}\right)^2}\)</td></tr><tr><td><strong>Clark's divergence distance</strong>                   <br><code>clark()</code></td><td>\(\displaystyle \sqrt{\sum_{i=1}^{n}\left(\frac{P_i - Q_i}{P_i + Q_i}\right)^{2}}\)</td></tr><tr><td><strong>Divergence</strong>                                    <br><code>divergence()</code></td><td>\(\displaystyle 2\sum_{i=1}^{n} \frac{(P_i - Q_i)^2}{(P_i + Q_i)^2}\)</td></tr><tr><td><strong>Euclidean distance</strong>                            <br><code>euclidean()</code></td><td>\(\sqrt{\sum_{i=1}^{n} (P_i - Q_i)^2}\)</td></tr><tr><td><strong>Gower distance</strong>                                <br><code>gower()</code></td><td>\(\displaystyle \frac{1}{n}\sum_{i=1}^{n}\frac{|P_i - Q_i|}{R_i}\)</td></tr><tr><td><strong>Hellinger distance</strong>                            <br><code>hellinger()</code></td><td>\(\sqrt{\sum_{i=1}^{n}(\sqrt{P_i} - \sqrt{Q_i})^{2}}\)</td></tr><tr><td><strong>Horn-Morisita dissimilarity</strong>                   <br><code>horn()</code></td><td>\(\displaystyle 1 - \frac{2\sum_{i=1}^{n}P_{i}Q_{i}}{\sum_{i=1}^{n}P_i^2 + \sum_{i=1}^{n}Q_i^2}\)</td></tr><tr><td><strong>Jensen-Shannon distance</strong>                       <br><code>jensen()</code></td><td>\(\displaystyle \sqrt{\frac{1}{2}\left[\sum_{i=1}^{n}P_i\ln\left(\frac{2P_i}{P_i + Q_i}\right) + \sum_{i=1}^{n}Q_i\ln\left(\frac{2Q_i}{P_i + Q_i}\right)\right]}\)</td></tr><tr><td><strong>Jensen-Shannon divergence (JSD)</strong>               <br><code>jsd()</code></td><td>\(\displaystyle \frac{1}{2}\left[\sum_{i=1}^{n}P_i\ln\left(\frac{2P_i}{P_i + Q_i}\right) + \sum_{i=1}^{n}Q_i\ln\left(\frac{2Q_i}{P_i + Q_i}\right)\right]\)</td></tr><tr><td><strong>Lorentzian distance</strong>                           <br><code>lorentzian()</code></td><td>\(\sum_{i=1}^{n}\ln{(1 + |P_i - Q_i|)}\)</td></tr><tr><td><strong>Manhattan distance</strong>                            <br><code>manhattan()</code></td><td>\(\sum_{i=1}^{n} |P_i - Q_i|\)</td></tr><tr><td><strong>Matusita distance</strong>                             <br><code>matusita()</code></td><td>\(\sqrt{\sum_{i=1}^{n}\left(\sqrt{P_i} - \sqrt{Q_i}\right)^2}\)</td></tr><tr><td><strong>Minkowski distance</strong>                            <br><code>minkowski()</code></td><td>\(\sqrt[p]{\sum_{i=1}^{n} (P_i - Q_i)^p}\) <br> Where \(p\) is the geometry of the space.</td></tr><tr><td><strong>Morisita dissimilarity</strong> <br> * Integers Only   <br><code>morisita()</code></td><td>\(\displaystyle 1 - \frac{2\sum_{i=1}^{n}X_{i}Y_{i}}{\displaystyle \left(\frac{\sum_{i=1}^{n}X_i(X_i - 1)}{X_T(X_T - 1)} + \frac{\sum_{i=1}^{n}Y_i(Y_i - 1)}{Y_T(Y_T - 1)}\right)X_{T}Y_{T}}\)</td></tr><tr><td><strong>Motyka dissimilarity</strong>                          <br><code>motyka()</code></td><td>\(\displaystyle \frac{\sum_{i=1}^{n} \max(P_i, Q_i)}{\sum_{i=1}^{n} (P_i + Q_i)}\)</td></tr><tr><td><strong>Probabilistic Symmetric \(\chi^2\) distance</strong> <br><code>psym_chisq()</code></td><td>\(\displaystyle 2\sum_{i=1}^{n}\frac{(P_i - Q_i)^2}{P_i + Q_i}\)</td></tr><tr><td><strong>Soergel distance</strong>                              <br><code>soergel()</code></td><td>\(\displaystyle \frac{\sum_{i=1}^{n} |P_i - Q_i|}{\sum_{i=1}^{n} \max(P_i, Q_i)}\)</td></tr><tr><td><strong>Squared \(\chi^2\) distance</strong>                 <br><code>squared_chisq()</code></td><td>\(\displaystyle \sum_{i=1}^{n}\frac{(P_i - Q_i)^2}{P_i + Q_i}\)</td></tr><tr><td><strong>Squared Chord distance</strong>                        <br><code>squared_chord()</code></td><td>\(\sum_{i=1}^{n}\left(\sqrt{P_i} - \sqrt{Q_i}\right)^2\)</td></tr><tr><td><strong>Squared Euclidean distance</strong>                    <br><code>squared_euclidean()</code></td><td>\(\sum_{i=1}^{n} (P_i - Q_i)^2\)</td></tr><tr><td><strong>Topsoe distance</strong>                               <br><code>topsoe()</code></td><td>\(\displaystyle \sum_{i=1}^{n}P_i\ln\left(\frac{2P_i}{P_i + Q_i}\right) + \sum_{i=1}^{n}Q_i\ln\left(\frac{2Q_i}{P_i + Q_i}\right)\)</td></tr><tr><td><strong>Wave Hedges distance</strong>                          <br><code>wave_hedges()</code></td><td>\(\displaystyle \frac{\sum_{i=1}^{n} |P_i - Q_i|}{\sum_{i=1}^{n} \max(P_i, Q_i)}\)</td></tr></table><div class="section">
<h3 id="presence-absence">Presence / Absence<a class="anchor" aria-label="anchor" href="#presence-absence"></a></h3>


<p>Given:</p><ul><li><p>\(A\), \(B\) : Number of features in each sample.</p></li>
<li><p>\(J\) : Number of features in common.</p></li>
</ul><table class="table table"><tr><td></td><td></td></tr><tr><td><strong>Dice-Sorensen dissimilarity</strong> <br><code>sorensen()</code></td><td>\(\displaystyle \frac{2J}{(A + B)}\)</td></tr><tr><td><strong>Hamming distance</strong>            <br><code>hamming()</code></td><td>\(\displaystyle (A + B) - 2J\)</td></tr><tr><td><strong>Jaccard distance</strong>            <br><code>jaccard()</code></td><td>\(\displaystyle 1 - \frac{J}{(A + B - J)]}\)</td></tr><tr><td><strong>Otsuka-Ochiai dissimilarity</strong> <br><code>ochiai()</code></td><td>\(\displaystyle 1 - \frac{J}{\sqrt{AB}}\)</td></tr></table></div>

<div class="section">
<h3 id="phylogenetic">Phylogenetic<a class="anchor" aria-label="anchor" href="#phylogenetic"></a></h3>


<p>Given \(n\) branches with lengths \(L\) and a pair of samples' binary
(\(A\) and \(B\)) or proportional abundances (\(P\) and \(Q\)) on
each of those branches.</p><table class="table table"><tr><td></td><td></td></tr><tr><td><strong>Unweighted UniFrac</strong>                  <br><code>unweighted_unifrac()</code></td><td>\(\displaystyle \frac{1}{n}\sum_{i=1}^{n} L_i|A_i - B_i|\)</td></tr><tr><td><strong>Weighted UniFrac</strong>                    <br><code>weighted_unifrac()</code></td><td>\(\displaystyle \sum_{i=1}^{n} L_i|P_i - Q_i|\)</td></tr><tr><td><strong>Normalized Weighted UniFrac</strong>         <br><code>normalized_unifrac()</code></td><td>\(\displaystyle \frac{\sum_{i=1}^{n} L_i|P_i - Q_i|}{\sum_{i=1}^{n} L_i(P_i + Q_i)}\)</td></tr><tr><td><strong>Generalized UniFrac (GUniFrac)</strong>      <br><code>generalized_unifrac()</code></td><td>\(\displaystyle \frac{\sum_{i=1}^{n} L_i(P_i + Q_i)^{\alpha}\left|\displaystyle \frac{P_i - Q_i}{P_i + Q_i}\right|}{\sum_{i=1}^{n} L_i(P_i + Q_i)^{\alpha}}\) <br> Where \(\alpha\) is a scalable weighting factor.</td></tr><tr><td><strong>Variance-Adjusted Weighted UniFrac</strong>  <br><code>variance_adjusted_unifrac()</code></td><td>\(\displaystyle \frac{\displaystyle \sum_{i=1}^{n} L_i\displaystyle \frac{|P_i - Q_i|}{\sqrt{(P_i + Q_i)(2 - P_i - Q_i)}} }{\displaystyle \sum_{i=1}^{n} L_i\displaystyle \frac{P_i + Q_i}{\sqrt{(P_i + Q_i)(2 - P_i - Q_i)}} }\)</td></tr></table><p>See <code><a href="../articles/unifrac.html">vignette('unifrac')</a></code> for detailed example UniFrac calculations.</p>
</div>

    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Levy, A., Shalom, B. R., &amp; Chalamish, M. (2024). A guide to similarity
measures. <em>arXiv</em>.</p>
<p>Cha, S.-H. (2007). Comprehensive survey on distance/similarity measures
between probability density functions. <em>International Journal of Mathematical
Models and Methods in Applied Sciences</em>, 1(4), 300–307.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="co"># Example counts matrix</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   Saliva Gums Nose Stool</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Streptococcus        162  793   22     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteroides            2    4    2   611</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Corynebacterium        0    0  498     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Haemophilus          180   87    2     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Propionibacterium      1    1  251     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Staphylococcus         0    1  236     1</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">bray</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4265973                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.9713843 0.9720256          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9909509 0.9911046 0.9915177</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">jaccard</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.2000000                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.3333333 0.1666667          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.5000000 0.3333333 0.1666667</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">generalized_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4471644                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.8215129 0.7607876          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9727827 0.9784242 0.9730332</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># Only calculate distances for Saliva vs all.</span></span></span>
<span class="r-in"><span>    <span class="fu">bray</span><span class="op">(</span><span class="va">ex_counts</span>, pairs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4265973                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.9713843        NA          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9909509        NA        NA</span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

