<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Beta Diversity Wrapper Function — beta_div • ecodive</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Beta Diversity Wrapper Function — beta_div"><meta name="description" content="Beta Diversity Wrapper Function"><meta property="og:description" content="Beta Diversity Wrapper Function"><meta property="og:image" content="https://cmmr.github.io/ecodive/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9003</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Beta Diversity Wrapper Function</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/main/R/beta_div.r" class="external-link"><code>R/beta_div.r</code></a></small>
      <div class="d-none name"><code>beta_div.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Beta Diversity Wrapper Function</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">beta_div</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  <span class="va">metric</span>,</span>
<span>  rescale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  power <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  pseudocount <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  tree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pairs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>A numeric matrix of count data where each column is a
feature, and each row is a sample. Any object coercible with
<code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code> can be given here, as well as <code>phyloseq</code>, <code>rbiom</code>,
<code>SummarizedExperiment</code>, and <code>TreeSummarizedExperiment</code> objects.</p></dd>


<dt id="arg-metric">metric<a class="anchor" aria-label="anchor" href="#arg-metric"></a></dt>
<dd><p>The name of a beta diversity metric. One of <code>c('aitchison', 'bhattacharyya', 'bray', 'canberra', 'chebyshev', 'chord', 'clark', 'divergence', 'euclidean', 'generalized_unifrac', 'gower', 'hamming', 'hellinger', 'horn', 'jaccard', 'jensen', 'jsd', 'lorentzian', 'manhattan', 'matusita', 'minkowski', 'morisita', 'motyka', 'normalized_unifrac', 'ochiai', 'psym_chisq', 'soergel', 'sorensen', 'squared_chisq', 'squared_chord', 'squared_euclidean', 'topsoe', 'unweighted_unifrac', 'variance_adjusted_unifrac', 'wave_hedges', 'weighted_unifrac')</code>. Flexible
matching is supported (see below). Programmatic access via
<code>list_metrics('beta')</code>.</p></dd>


<dt id="arg-rescale">rescale<a class="anchor" aria-label="anchor" href="#arg-rescale"></a></dt>
<dd><p>Normalize each sample's counts so they sum to <code>1</code>.
Default: <code>TRUE</code></p></dd>


<dt id="arg-power">power<a class="anchor" aria-label="anchor" href="#arg-power"></a></dt>
<dd><p>Scaling factor for the magnitude of differences between
communities (\(p\)). Default: <code>1.5</code></p></dd>


<dt id="arg-pseudocount">pseudocount<a class="anchor" aria-label="anchor" href="#arg-pseudocount"></a></dt>
<dd><p>The value to add to all counts in <code>counts</code> to prevent
taking <code>log(0)</code> for unobserved features. The default, <code>NULL</code>, selects
the smallest non-zero value in <code>counts</code>.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>How much weight to give to relative abundances; a value
between 0 and 1, inclusive. Setting <code>alpha=1</code> is equivalent to
<code><a href="bdiv_functions.html">normalized_unifrac()</a></code>.</p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>A <code>phylo</code>-class object representing the phylogenetic tree for
the OTUs in <code>counts</code>. The OTU identifiers given by <code>colnames(counts)</code>
must be present in <code>tree</code>. Can be omitted if a tree is embedded with
the <code>counts</code> object or as <code>attr(counts, 'tree')</code>.</p></dd>


<dt id="arg-pairs">pairs<a class="anchor" aria-label="anchor" href="#arg-pairs"></a></dt>
<dd><p>Which combinations of samples should distances be
calculated for? The default value (<code>NULL</code>) calculates all-vs-all.
Provide a numeric or logical vector specifying positions in the
distance matrix to calculate. See examples.</p></dd>


<dt id="arg-cpus">cpus<a class="anchor" aria-label="anchor" href="#arg-cpus"></a></dt>
<dd><p>How many parallel processing threads should be used. The
default, <code><a href="n_cpus.html">n_cpus()</a></code>, will use all logical CPU cores.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional options to pass through to the called function. I.e.
<code>tree</code>, <code>pairs</code>, <code>alpha</code>, or <code>cpus</code>. Named options will only be passed
though if the function expects them.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A numeric vector.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><strong>List of Beta Diversity Metrics</strong></p><table class="table table"><tr><td>Option / Function Name</td><td>Metric Name</td></tr><tr><td><code>aitchison</code></td><td>Aitchison distance</td></tr><tr><td><code>bhattacharyya</code></td><td>Bhattacharyya distance</td></tr><tr><td><code>bray</code></td><td>Bray-Curtis dissimilarity</td></tr><tr><td><code>canberra</code></td><td>Canberra distance</td></tr><tr><td><code>chebyshev</code></td><td>Chebyshev distance</td></tr><tr><td><code>chord</code></td><td>Chord distance</td></tr><tr><td><code>clark</code></td><td>Clark's divergence distance</td></tr><tr><td><code>divergence</code></td><td>Divergence</td></tr><tr><td><code>euclidean</code></td><td>Euclidean distance</td></tr><tr><td><code>generalized_unifrac</code></td><td>Generalized UniFrac (GUniFrac)</td></tr><tr><td><code>gower</code></td><td>Gower distance</td></tr><tr><td><code>hamming</code></td><td>Hamming distance</td></tr><tr><td><code>hellinger</code></td><td>Hellinger distance</td></tr><tr><td><code>horn</code></td><td>Horn-Morisita dissimilarity</td></tr><tr><td><code>jaccard</code></td><td>Jaccard distance</td></tr><tr><td><code>jensen</code></td><td>Jensen-Shannon distance</td></tr><tr><td><code>jsd</code></td><td>Jesen-Shannon divergence (JSD)</td></tr><tr><td><code>lorentzian</code></td><td>Lorentzian distance</td></tr><tr><td><code>manhattan</code></td><td>Manhattan distance</td></tr><tr><td><code>matusita</code></td><td>Matusita distance</td></tr><tr><td><code>minkowski</code></td><td>Minkowski distance</td></tr><tr><td><code>morisita</code></td><td>Morisita dissimilarity</td></tr><tr><td><code>motyka</code></td><td>Motyka dissimilarity</td></tr><tr><td><code>normalized_unifrac</code></td><td>Normalized Weighted UniFrac</td></tr><tr><td><code>ochiai</code></td><td>Otsuka-Ochiai dissimilarity</td></tr><tr><td><code>psym_chisq</code></td><td>Probabilistic Symmetric Chi-Squared distance</td></tr><tr><td><code>soergel</code></td><td>Soergel distance</td></tr><tr><td><code>sorensen</code></td><td>Dice-Sorensen dissimilarity</td></tr><tr><td><code>squared_chisq</code></td><td>Squared Chi-Squared distance</td></tr><tr><td><code>squared_chord</code></td><td>Squared Chord distance</td></tr><tr><td><code>squared_euclidean</code></td><td>Squared Euclidean distance</td></tr><tr><td><code>topsoe</code></td><td>Topsoe distance</td></tr><tr><td><code>unweighted_unifrac</code></td><td>Unweighted UniFrac</td></tr><tr><td><code>variance_adjusted_unifrac</code></td><td>Variance-Adjusted Weighted UniFrac (VAW-UniFrac)</td></tr><tr><td><code>wave_hedges</code></td><td>Wave Hedges distance</td></tr><tr><td><code>weighted_unifrac</code></td><td>Weighted UniFrac</td></tr></table><p><strong>Flexible name matching</strong></p>
<p>Case insensitive and partial matching. Any runs of non-alpha characters are
converted to underscores. E.g. <code>metric = 'Weighted UniFrac</code> selects
<code>weighted_unifrac</code>.</p>
<p>UniFrac names can be shortened to the first letter plus "unifrac". E.g.
<code>uunifrac</code>, <code>w_unifrac</code>, or <code>V UniFrac</code>. These also support partial matching.</p>
<p>Finished code should always use the full primary option name to avoid
ambiguity with future additions to the metrics list.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="co"># Example counts matrix</span></span></span>
<span class="r-in"><span>    <span class="va">ex_counts</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Streptococcus Bacteroides Corynebacterium Haemophilus Propionibacterium</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Saliva           162           2               0         180                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums             793           4               0          87                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose              22           2             498           2               251</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool              1         611               1           1                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Staphylococcus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Saliva              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums                1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose              236</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool               1</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># Bray-Curtis distances</span></span></span>
<span class="r-in"><span>    <span class="fu">beta_div</span><span class="op">(</span><span class="va">ex_counts</span>, <span class="st">'bray'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4265973                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.9713843 0.9720256          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9909509 0.9911046 0.9915177</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># Generalized UniFrac distances</span></span></span>
<span class="r-in"><span>    <span class="fu">beta_div</span><span class="op">(</span><span class="va">ex_counts</span>, <span class="st">'GUniFrac'</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4471644                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.8215129 0.7607876          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9727827 0.9784242 0.9730332</span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

