<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • ecodive</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"><meta property="og:image" content="https://cmmr.github.io/ecodive/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/performance.html">Performance Guide</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="alpha-diversity">Alpha Diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a></h2>

      <div class="section-desc"><p>Calculate within-sample diversity using various indices.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="alpha_div.html">alpha_div()</a></code>

        </dt>
        <dd>Alpha Diversity Wrapper Function</dd>
      </dl></div><div class="section level2">

      <h3 id="richness-metrics">Richness metrics<a class="anchor" aria-label="anchor" href="#richness-metrics"></a></h3>
      <div class="section-desc"><p>Estimators focused on the number of species (richness).</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ace.html">ace()</a></code>

        </dt>
        <dd>Abundance-based Coverage Estimator (ACE)</dd>

        <dt>

          <code><a href="chao1.html">chao1()</a></code>

        </dt>
        <dd>Chao1 Richness Estimator</dd>

        <dt>

          <code><a href="margalef.html">margalef()</a></code>

        </dt>
        <dd>Margalef's Richness Index</dd>

        <dt>

          <code><a href="menhinick.html">menhinick()</a></code>

        </dt>
        <dd>Menhinick's Richness Index</dd>

        <dt>

          <code><a href="observed.html">observed()</a></code>

        </dt>
        <dd>Observed Features</dd>

        <dt>

          <code><a href="squares.html">squares()</a></code>

        </dt>
        <dd>Squares Richness Estimator</dd>
      </dl></div><div class="section level2">

      <h3 id="diversity-metrics">Diversity metrics<a class="anchor" aria-label="anchor" href="#diversity-metrics"></a></h3>
      <div class="section-desc"><p>Indices accounting for both richness and evenness.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="brillouin.html">brillouin()</a></code>

        </dt>
        <dd>Brillouin Index</dd>

        <dt>

          <code><a href="fisher.html">fisher()</a></code>

        </dt>
        <dd>Fisher's Alpha</dd>

        <dt>

          <code><a href="inv_simpson.html">inv_simpson()</a></code>

        </dt>
        <dd>Inverse Simpson Index</dd>

        <dt>

          <code><a href="shannon.html">shannon()</a></code>

        </dt>
        <dd>Shannon Diversity Index</dd>

        <dt>

          <code><a href="simpson.html">simpson()</a></code>

        </dt>
        <dd>Gini-Simpson Index</dd>
      </dl></div><div class="section level2">

      <h3 id="dominance-metrics">Dominance metrics<a class="anchor" aria-label="anchor" href="#dominance-metrics"></a></h3>
      <div class="section-desc"><p>Indices focused on the abundance of the most dominant species.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="berger.html">berger()</a></code>

        </dt>
        <dd>Berger-Parker Index</dd>

        <dt>

          <code><a href="mcintosh.html">mcintosh()</a></code>

        </dt>
        <dd>McIntosh Index</dd>
      </dl></div><div class="section level2">

      <h3 id="phylogenetic-metrics">Phylogenetic metrics<a class="anchor" aria-label="anchor" href="#phylogenetic-metrics"></a></h3>
      <div class="section-desc"><p>Diversity based on phylogenetic branch lengths.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="faith.html">faith()</a></code>

        </dt>
        <dd>Faith's Phylogenetic Diversity (PD)</dd>
      </dl></div><div class="section level2">
      <h2 id="beta-diversity">Beta Diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a></h2>

      <div class="section-desc"><p>Calculate between-sample similarity or distance.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="beta_div.html">beta_div()</a></code>

        </dt>
        <dd>Beta Diversity Wrapper Function</dd>
      </dl></div><div class="section level2">

      <h3 id="abundance-metrics">Abundance metrics<a class="anchor" aria-label="anchor" href="#abundance-metrics"></a></h3>
      <div class="section-desc"><p>Distance measures using quantitative abundance data.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="aitchison.html">aitchison()</a></code>

        </dt>
        <dd>Aitchison distance</dd>

        <dt>

          <code><a href="bhattacharyya.html">bhattacharyya()</a></code>

        </dt>
        <dd>Bhattacharyya distance</dd>

        <dt>

          <code><a href="bray.html">bray()</a></code>

        </dt>
        <dd>Bray-Curtis dissimilarity</dd>

        <dt>

          <code><a href="canberra.html">canberra()</a></code>

        </dt>
        <dd>Canberra distance</dd>

        <dt>

          <code><a href="chebyshev.html">chebyshev()</a></code>

        </dt>
        <dd>Chebyshev distance</dd>

        <dt>

          <code><a href="chord.html">chord()</a></code>

        </dt>
        <dd>Chord distance</dd>

        <dt>

          <code><a href="clark.html">clark()</a></code>

        </dt>
        <dd>Clark's divergence distance</dd>

        <dt>

          <code><a href="divergence.html">divergence()</a></code>

        </dt>
        <dd>Divergence</dd>

        <dt>

          <code><a href="euclidean.html">euclidean()</a></code>

        </dt>
        <dd>Euclidean distance</dd>

        <dt>

          <code><a href="gower.html">gower()</a></code>

        </dt>
        <dd>Gower distance</dd>

        <dt>

          <code><a href="hellinger.html">hellinger()</a></code>

        </dt>
        <dd>Hellinger distance</dd>

        <dt>

          <code><a href="horn.html">horn()</a></code>

        </dt>
        <dd>Horn-Morisita dissimilarity</dd>

        <dt>

          <code><a href="jensen.html">jensen()</a></code>

        </dt>
        <dd>Jensen-Shannon distance</dd>

        <dt>

          <code><a href="jsd.html">jsd()</a></code>

        </dt>
        <dd>Jensen-Shannon divergence (JSD)</dd>

        <dt>

          <code><a href="lorentzian.html">lorentzian()</a></code>

        </dt>
        <dd>Lorentzian distance</dd>

        <dt>

          <code><a href="manhattan.html">manhattan()</a></code>

        </dt>
        <dd>Manhattan distance</dd>

        <dt>

          <code><a href="matusita.html">matusita()</a></code>

        </dt>
        <dd>Matusita distance</dd>

        <dt>

          <code><a href="minkowski.html">minkowski()</a></code>

        </dt>
        <dd>Minkowski distance</dd>

        <dt>

          <code><a href="morisita.html">morisita()</a></code>

        </dt>
        <dd>Morisita dissimilarity</dd>

        <dt>

          <code><a href="motyka.html">motyka()</a></code>

        </dt>
        <dd>Motyka dissimilarity</dd>

        <dt>

          <code><a href="psym_chisq.html">psym_chisq()</a></code>

        </dt>
        <dd>Probabilistic Symmetric Chi-Squared distance</dd>

        <dt>

          <code><a href="soergel.html">soergel()</a></code>

        </dt>
        <dd>Soergel distance</dd>

        <dt>

          <code><a href="squared_chisq.html">squared_chisq()</a></code>

        </dt>
        <dd>Squared Chi-Squared distance</dd>

        <dt>

          <code><a href="squared_chord.html">squared_chord()</a></code>

        </dt>
        <dd>Squared Chord distance</dd>

        <dt>

          <code><a href="squared_euclidean.html">squared_euclidean()</a></code>

        </dt>
        <dd>Squared Euclidean distance</dd>

        <dt>

          <code><a href="topsoe.html">topsoe()</a></code>

        </dt>
        <dd>Topsoe distance</dd>

        <dt>

          <code><a href="wave_hedges.html">wave_hedges()</a></code>

        </dt>
        <dd>Wave Hedges distance</dd>
      </dl></div><div class="section level2">

      <h3 id="presence-absence-metrics">Presence/Absence metrics<a class="anchor" aria-label="anchor" href="#presence-absence-metrics"></a></h3>
      <div class="section-desc"><p>Distance measures using binary (presence/absence) data.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="hamming.html">hamming()</a></code>

        </dt>
        <dd>Hamming distance</dd>

        <dt>

          <code><a href="jaccard.html">jaccard()</a></code>

        </dt>
        <dd>Jaccard distance</dd>

        <dt>

          <code><a href="ochiai.html">ochiai()</a></code>

        </dt>
        <dd>Otsuka-Ochiai dissimilarity</dd>

        <dt>

          <code><a href="sorensen.html">sorensen()</a></code>

        </dt>
        <dd>Dice-Sorensen dissimilarity</dd>
      </dl></div><div class="section level2">

      <h3 id="phylogenetic-metrics-1">Phylogenetic metrics<a class="anchor" aria-label="anchor" href="#phylogenetic-metrics-1"></a></h3>
      <div class="section-desc"><p>Distance measures based on phylogenetic branch lengths (UniFrac family).</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="unweighted_unifrac.html">unweighted_unifrac()</a></code>

        </dt>
        <dd>Unweighted UniFrac</dd>

        <dt>

          <code><a href="weighted_unifrac.html">weighted_unifrac()</a></code>

        </dt>
        <dd>Weighted UniFrac</dd>

        <dt>

          <code><a href="normalized_unifrac.html">normalized_unifrac()</a></code>

        </dt>
        <dd>Normalized Weighted UniFrac</dd>

        <dt>

          <code><a href="generalized_unifrac.html">generalized_unifrac()</a></code>

        </dt>
        <dd>Generalized UniFrac (GUniFrac)</dd>

        <dt>

          <code><a href="variance_adjusted_unifrac.html">variance_adjusted_unifrac()</a></code>

        </dt>
        <dd>Variance-Adjusted Weighted UniFrac</dd>
      </dl></div><div class="section level2">
      <h2 id="utilities-helpers">Utilities &amp; Helpers<a class="anchor" aria-label="anchor" href="#utilities-helpers"></a></h2>

      <div class="section-desc"><p>Tools for programmatic access, data manipulation, and configuration.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="metrics.html">list_metrics()</a></code> <code><a href="metrics.html">match_metric()</a></code>

        </dt>
        <dd>Find and Browse Available Metrics</dd>

        <dt>

          <code><a href="read_tree.html">read_tree()</a></code>

        </dt>
        <dd>Read a newick formatted phylogenetic tree.</dd>

        <dt>

          <code><a href="rarefy.html">rarefy()</a></code>

        </dt>
        <dd>Rarefy Observation Counts</dd>

        <dt>

          <code><a href="n_cpus.html">n_cpus()</a></code>

        </dt>
        <dd>Number of CPU Cores</dd>
      </dl></div><div class="section level2">
      <h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a></h2>

      <div class="section-desc"><p>Example data provided with the package.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ex_counts.html">ex_counts</a></code>

        </dt>
        <dd>Example counts matrix</dd>

        <dt>

          <code><a href="ex_tree.html">ex_tree</a></code>

        </dt>
        <dd>Example phylogenetic tree</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

