<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Normalized UniFrac beta diversity metric — normalized_unifrac • fastbiom</title><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Normalized UniFrac beta diversity metric — normalized_unifrac"><meta name="description" content="Normalized UniFrac beta diversity metric"><meta property="og:description" content="Normalized UniFrac beta diversity metric"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fastbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/fastbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Normalized UniFrac beta diversity metric</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/fastbiom/blob/main/R/beta_div.r" class="external-link"><code>R/beta_div.r</code></a></small>
      <div class="d-none name"><code>normalized_unifrac.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Normalized UniFrac beta diversity metric</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">normalized_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>An OTU abundance matrix where each column is a sample, and
each row is an OTU. Any object coercible with <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code> can be
given here, as well as <code>phyloseq</code>, <code>rbiom</code>, <code>SummarizedExperiment</code>,
and <code>TreeSummarizedExperiment</code> objects.</p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>A <code>phylo</code>-class object representing the phylogenetic tree for
the OTUs in <code>counts</code>. The OTU identifiers given by <code>colnames(counts)</code>
must be present in <code>tree</code>. Can be omitted if a tree is embedded with
the <code>counts</code> object or as <code>attr(counts, 'tree')</code>.</p></dd>


<dt id="arg-pairs">pairs<a class="anchor" aria-label="anchor" href="#arg-pairs"></a></dt>
<dd><p>Which combinations of samples should distances be
calculated for? The default value (<code>NULL</code>) calculates all-vs-all.
Provide a numeric or logical vector specifying positions in the
distance matrix to calculate. See examples.</p></dd>


<dt id="arg-cpus">cpus<a class="anchor" aria-label="anchor" href="#arg-cpus"></a></dt>
<dd><p>How many parallel processing threads should be used. The
default, <code><a href="n_cpus.html">n_cpus()</a></code>, will use all logical CPU cores.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="calculation">Calculation<a class="anchor" aria-label="anchor" href="#calculation"></a></h2>



<p>Given \(n\) branches with lengths \(L\) and a pair of samples'
abundances (\(A\) and \(B\)) on each of those branches:</p>
<p>$$\displaystyle \frac{\sum_{i = 1}^{n} L_i|\frac{A_i}{A_T} - \frac{B_i}{B_T}|}{\sum_{i = 1}^{n} L_i(\frac{A_i}{A_T} + \frac{B_i}{B_T})}$$</p>
<p>See <a href="https://cmmr.github.io/fastbiom/articles/unifrac.html">https://cmmr.github.io/fastbiom/articles/unifrac.html</a> for details and
a worked example.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Lozupone CA, Hamady M, Kelley ST, Knight R 2007.
Quantitative and Qualitative \(\beta\) Diversity Measures Lead to Different Insights into Factors That Structure Microbial Communities.
Applied and Environmental Microbiology, 73(5).
<a href="https://doi.org/10.1128/AEM.01996-06" class="external-link">https://doi.org/10.1128/AEM.01996-06</a></p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other beta_diversity:
<code><a href="bray_curtis.html">bray_curtis</a>()</code>,
<code><a href="canberra.html">canberra</a>()</code>,
<code><a href="euclidean.html">euclidean</a>()</code>,
<code><a href="generalized_unifrac.html">generalized_unifrac</a>()</code>,
<code><a href="gower.html">gower</a>()</code>,
<code><a href="jaccard.html">jaccard</a>()</code>,
<code><a href="kulczynski.html">kulczynski</a>()</code>,
<code><a href="manhattan.html">manhattan</a>()</code>,
<code><a href="unweighted_unifrac.html">unweighted_unifrac</a>()</code>,
<code><a href="variance_adjusted_unifrac.html">variance_adjusted_unifrac</a>()</code>,
<code><a href="weighted_unifrac.html">weighted_unifrac</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="co"># Example counts matrix</span></span></span>
<span class="r-in"><span>    <span class="va">ex_counts</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      A B C  D</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> OTU1 4 0 0  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> OTU2 0 8 9 10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> OTU3 3 0 0  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> OTU4 2 0 0  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> OTU5 6 5 7  1</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># UniFrac weighted distance matrix</span></span></span>
<span class="r-in"><span>    <span class="fu">normalized_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            A          B          C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B 0.51102941                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> C 0.49654112 0.04820717           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> D 0.69136961 0.22837370 0.27567790</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># Only calculate distances for A vs all.</span></span></span>
<span class="r-in"><span>    <span class="fu">normalized_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span>, pairs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           A         B         C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B 0.5110294                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> C 0.4965411        NA          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> D 0.6913696        NA        NA</span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

