<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Alpha Diversity Metrics — adiv_functions • ecodive</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Alpha Diversity Metrics — adiv_functions"><meta name="description" content="Alpha Diversity Metrics"><meta property="og:description" content="Alpha Diversity Metrics"><meta property="og:image" content="https://cmmr.github.io/ecodive/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0.9006</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Alpha Diversity Metrics</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/HEAD/R/adiv_functions.r" class="external-link"><code>R/adiv_functions.r</code></a></small>
      <div class="d-none name"><code>adiv_functions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Alpha Diversity Metrics</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ace</span><span class="op">(</span><span class="va">x</span>, cutoff <span class="op">=</span> <span class="fl">10</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">berger</span><span class="op">(</span><span class="va">x</span>, rescale <span class="op">=</span> <span class="cn">TRUE</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">brillouin</span><span class="op">(</span><span class="va">x</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">chao1</span><span class="op">(</span><span class="va">x</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">faith</span><span class="op">(</span><span class="va">x</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">fisher</span><span class="op">(</span><span class="va">x</span>, digits <span class="op">=</span> <span class="fl">3L</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">inv_simpson</span><span class="op">(</span><span class="va">x</span>, rescale <span class="op">=</span> <span class="cn">TRUE</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">margalef</span><span class="op">(</span><span class="va">x</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mcintosh</span><span class="op">(</span><span class="va">x</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">menhinick</span><span class="op">(</span><span class="va">x</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">observed</span><span class="op">(</span><span class="va">x</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">shannon</span><span class="op">(</span><span class="va">x</span>, rescale <span class="op">=</span> <span class="cn">TRUE</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">simpson</span><span class="op">(</span><span class="va">x</span>, rescale <span class="op">=</span> <span class="cn">TRUE</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">squares</span><span class="op">(</span><span class="va">x</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A numeric matrix of count data where each column is a feature,
and each row is a sample. Any object coercible with <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code>
can be given here, as well as <code>phyloseq</code>, <code>rbiom</code>,
<code>SummarizedExperiment</code>, and <code>TreeSummarizedExperiment</code> objects.</p></dd>


<dt id="arg-cutoff">cutoff<a class="anchor" aria-label="anchor" href="#arg-cutoff"></a></dt>
<dd><p>The maximum number of observations to consider "rare".
Default: <code>10</code>.</p></dd>


<dt id="arg-cpus">cpus<a class="anchor" aria-label="anchor" href="#arg-cpus"></a></dt>
<dd><p>How many parallel processing threads should be used. The
default, <code><a href="n_cpus.html">n_cpus()</a></code>, will use all logical CPU cores.</p></dd>


<dt id="arg-rescale">rescale<a class="anchor" aria-label="anchor" href="#arg-rescale"></a></dt>
<dd><p>Normalize each sample's counts so they sum to <code>1</code>.
Default: <code>TRUE</code></p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>A <code>phylo</code>-class object representing the phylogenetic tree for
the OTUs in <code>x</code>. The OTU identifiers given by <code>colnames(x)</code>
must be present in <code>tree</code>. Can be omitted if a tree is embedded with
the <code>x</code> object or as <code>attr(x, 'tree')</code>.</p></dd>


<dt id="arg-digits">digits<a class="anchor" aria-label="anchor" href="#arg-digits"></a></dt>
<dd><p>Precision of the returned values, in number of decimal
places. E.g. the default <code>digits=3</code> could return <code>6.392</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A numeric vector.</p>
    </div>
    <div class="section level2">
    <h2 id="formulas">Formulas<a class="anchor" aria-label="anchor" href="#formulas"></a></h2>



<p>Prerequisite: all counts are whole numbers.</p>
<p>Given:</p><ul><li><p>\(n\) : The number of features (e.g. species, OTUs, ASVs, etc).</p></li>
<li><p>\(X_i\) : Integer count of the \(i\)-th feature.</p></li>
<li><p>\(X_T\) : Total of all counts (i.e. sequencing depth). \(X_T = \sum_{i=1}^{n} X_i\)</p></li>
<li><p>\(P_i\) : Proportional abundance of the \(i\)-th feature. \(P_i = X_i / X_T\)</p></li>
<li><p>\(F_1\) : Number of features where \(X_i = 1\) (i.e. singletons).</p></li>
<li><p>\(F_2\) : Number of features where \(X_i = 2\) (i.e. doubletons).</p></li>
</ul><table class="table table"><tr><td></td><td></td></tr><tr><td><strong>Abundance-based Coverage Estimator (ACE)</strong> <br><code>ace()</code></td><td>See below.</td></tr><tr><td><strong>Berger-Parker Index</strong>                      <br><code>berger()</code></td><td>\(\max(P_i)\)</td></tr><tr><td><strong>Brillouin Index</strong>                          <br><code>brillouin()</code></td><td>\(\displaystyle \frac{\ln{[(\sum_{i = 1}^{n} X_i)!]} - \sum_{i = 1}^{n} \ln{(X_i!)}}{\sum_{i = 1}^{n} X_i}\)</td></tr><tr><td><strong>Chao1</strong>                                    <br><code>chao1()</code></td><td>\(\displaystyle n + \frac{(F_1)^2}{2 F_2}\)</td></tr><tr><td><strong>Faith's Phylogenetic Diversity</strong>           <br><code>faith()</code></td><td>See below.</td></tr><tr><td><strong>Fisher's Alpha (\(\alpha\))</strong>            <br><code>fisher()</code></td><td>\(\displaystyle \frac{n}{\alpha} = \ln{\left(1 + \frac{X_T}{\alpha}\right)}\) <br> The value of \(\alpha\) must be solved for iteratively.</td></tr><tr><td><strong>Gini-Simpson Index</strong>                       <br><code>simpson()</code></td><td>\(1 - \sum_{i = 1}^{n} P_i^2\)</td></tr><tr><td><strong>Inverse Simpson Index</strong>                    <br><code>inv_simpson()</code></td><td>\(1 / \sum_{i = 1}^{n} P_i^2\)</td></tr><tr><td><strong>Margalef's Richness Index</strong>                <br><code>margalef()</code></td><td>\(\displaystyle \frac{n - 1}{\ln{X_T}}\)</td></tr><tr><td><strong>McIntosh Index</strong>                           <br><code>mcintosh()</code></td><td>\(\displaystyle \frac{X_T - \sqrt{\sum_{i = 1}^{n} (X_i)^2}}{X_T - \sqrt{X_T}}\)</td></tr><tr><td><strong>Menhinick's Richness Index</strong>               <br><code>menhinick()</code></td><td>\(\displaystyle \frac{n}{\sqrt{X_T}}\)</td></tr><tr><td><strong>Observed Features</strong>                        <br><code>observed()</code></td><td>\(n\)</td></tr><tr><td><strong>Shannon Diversity Index</strong>                  <br><code>shannon()</code></td><td>\(-\sum_{i = 1}^{n} P_i \times \ln(P_i)\)</td></tr><tr><td><strong>Squares Richness Estimator</strong>               <br><code>squares()</code></td><td>\(\displaystyle n + \frac{(F_1)^2 \sum_{i=1}^{n} (X_i)^2}{X_T^2 - nF_1}\)</td></tr></table><div class="section">
<h3 id="abundance-based-coverage-estimator-ace-">Abundance-based Coverage Estimator (ACE)<a class="anchor" aria-label="anchor" href="#abundance-based-coverage-estimator-ace-"></a></h3>


<p>Given:</p><ul><li><p>\(n\) : The number of features (e.g. species, OTUs, ASVs, etc).</p></li>
<li><p>\(r\) : Rare cutoff. Features with \(\le r\) counts are considered rare.</p></li>
<li><p>\(X_i\) : Integer count of the \(i\)-th feature.</p></li>
<li><p>\(F_i\) : Number of features with exactly \(i\) counts.</p></li>
<li><p>\(F_1\) : Number of features where \(X_i = 1\) (i.e. singletons).</p></li>
<li><p>\(F_{rare}\) : Number of rare features where \(X_i \le r\).</p></li>
<li><p>\(F_{abund}\) : Number of abundant features where \(X_i &gt; r\).</p></li>
<li><p>\(X_{rare}\) : Total counts belonging to rare features.</p></li>
<li><p>\(C_{ace}\) : The sample abundance coverage estimator, defined below.</p></li>
<li><p>\(\gamma_{ace}^2\) : The estimated coefficient of variation, defined below.</p></li>
<li><p>\(D_{ace}\) : Estimated number of features in the sample.</p></li>
</ul><p>\(\displaystyle C_{ace} = 1 - \frac{F_1}{X_{rare}}\)</p>
<p>\(\displaystyle \gamma_{ace}^2 = \max\left[\frac{F_{rare} \sum_{i=1}^{r}i(i-1)F_i}{C_{ace}X_{rare}(X_{rare} - 1)} - 1, 0\right]\)</p>
<p>\(\displaystyle D_{ace} = F_{abund} + \frac{F_{rare}}{C_{ace}} + \frac{F_1}{C_{ace}}\gamma_{ace}^2 \)</p>
</div>

<div class="section">
<h3 id="faith-s-phylogenetic-diversity-faith-s-pd-">Faith's Phylogenetic Diversity (Faith's PD)<a class="anchor" aria-label="anchor" href="#faith-s-phylogenetic-diversity-faith-s-pd-"></a></h3>


<p>Given \(n\) branches with lengths \(L\) and a sample's abundances
\(A\) on each of those branches coded as 1 for present or 0 for absent:</p>
<p>\(\sum_{i = 1}^{n} L_i A_i\)</p>
</div>

    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="co"># Example counts matrix</span></span></span>
<span class="r-in"><span>    <span class="va">ex_counts</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Streptococcus Bacteroides Corynebacterium Haemophilus Propionibacterium</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Saliva           162           2               0         180                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums             793           4               0          87                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose              22           2             498           2               251</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool              1         611               1           1                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Staphylococcus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Saliva              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums                1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose              236</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool               1</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">ace</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Saliva   Gums   Nose  Stool </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5.0    8.9    6.0    NaN </span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">chao1</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Saliva   Gums   Nose  Stool </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4.5    Inf    6.0    Inf </span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">squares</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Saliva      Gums      Nose     Stool </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  4.492762  8.243044  6.000000 20.793551 </span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

