<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Geometric Distances in Feature Space — bhattacharyya • ecodive</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Geometric Distances in Feature Space — bhattacharyya"><meta name="description" content="This large and diverse family of metrics shares a common conceptual
foundation: they treat each sample as a point in a multi-dimensional space,
where each species represents an axis and its abundance is the coordinate on
that axis. The dissimilarity between two samples is then calculated as the
geometric distance between the two points."><meta property="og:description" content="This large and diverse family of metrics shares a common conceptual
foundation: they treat each sample as a point in a multi-dimensional space,
where each species represents an axis and its abundance is the coordinate on
that axis. The dissimilarity between two samples is then calculated as the
geometric distance between the two points."><meta property="og:image" content="https://cmmr.github.io/ecodive/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0.9002</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Geometric Distances in Feature Space</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/HEAD/R/bdiv_geometric.r" class="external-link"><code>R/bdiv_geometric.r</code></a></small>
      <div class="d-none name"><code>bdiv_geometric.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This large and diverse family of metrics shares a common conceptual
foundation: they treat each sample as a point in a multi-dimensional space,
where each species represents an axis and its abundance is the coordinate on
that axis. The dissimilarity between two samples is then calculated as the
geometric distance between the two points.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bhattacharyya</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">canberra</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">chebyshev</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">chord</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">clark</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gower</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">euclidean</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hellinger</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lorentzian</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">manhattan</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">matusita</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">minkowski</span><span class="op">(</span><span class="va">counts</span>, power <span class="op">=</span> <span class="fl">1.5</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">squared_chord</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">squared_euclidean</span><span class="op">(</span><span class="va">counts</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>A numeric matrix of count data where each column is a sample,
and each row is a feature. Any object coercible with <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code>
can be given here, as well as <code>phyloseq</code>, <code>rbiom</code>,
<code>SummarizedExperiment</code>, and <code>TreeSummarizedExperiment</code> objects.</p></dd>


<dt id="arg-pairs">pairs<a class="anchor" aria-label="anchor" href="#arg-pairs"></a></dt>
<dd><p>Which combinations of samples should distances be
calculated for? The default value (<code>NULL</code>) calculates all-vs-all.
Provide a numeric or logical vector specifying positions in the
distance matrix to calculate. See examples.</p></dd>


<dt id="arg-cpus">cpus<a class="anchor" aria-label="anchor" href="#arg-cpus"></a></dt>
<dd><p>How many parallel processing threads should be used. The
default, <code><a href="n_cpus.html">n_cpus()</a></code>, will use all logical CPU cores.</p></dd>


<dt id="arg-power">power<a class="anchor" aria-label="anchor" href="#arg-power"></a></dt>
<dd><p>Scaling factor for the magnitude of differences between
communities (\(p\)). Default: <code>1.5</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>dist</code> object.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<div class="section">
<h3 id="the-minkowski-family-euclidean-manhattan-and-chebyshev">The Minkowski Family: Euclidean, Manhattan, and Chebyshev<a class="anchor" aria-label="anchor" href="#the-minkowski-family-euclidean-manhattan-and-chebyshev"></a></h3>


<p>The <strong>Minkowski distance</strong> is a generalized metric that encompasses several
common distances as special cases. Its formula is:</p>
<p>$$D_{Minkowski} =  \sqrt[p]{\sum_{i=1}^{n} (x_i - y_i)^p}$$</p>
<p>The parameter p controls the geometry of the space.</p><ul><li><p>When \(p = 2\), it becomes the <strong>Euclidean distance</strong>, the familiar "straight-line" distance between two points. In ecology, the raw Euclidean distance is highly problematic because it is sensitive to the "double zero problem"—the shared absence of a species contributes nothing to the distance, but it is also sensitive to total abundance and can give misleading results when species are on different scales. The <strong>Squared Euclidean distance</strong> simply omits the final square root, which places even greater weight on large differences.</p></li>
<li><p>When \(p = 1\), it becomes the <strong>Manhattan distance</strong> (also known as City Block distance), which is the sum of the absolute differences along each axis. It is less sensitive to outliers than Euclidean distance. The Bray-Curtis dissimilarity is a normalized version of the Manhattan distance.</p></li>
<li><p>When \(p \rightarrow \infty\), it becomes the <strong>Chebyshev distance</strong>, which is simply the maximum absolute difference along any single axis. This metric is useful when a single large difference is more important than the sum of many small differences, such as in anomaly detection.</p></li>
</ul></div>

<div class="section">
<h3 id="weighted-and-normalized-geometric-distances">Weighted and Normalized Geometric Distances<a class="anchor" aria-label="anchor" href="#weighted-and-normalized-geometric-distances"></a></h3>


<p>Several metrics modify the basic geometric distances to make them more suitable for ecological data.</p><ul><li><p>The <strong>Canberra distance</strong> is a weighted version of the Manhattan distance that is particularly sensitive to changes in values near zero. It divides the absolute difference for each species by the sum of their abundances, giving more weight to differences in rare species. However, this sensitivity can also amplify noise in sparse, low-count data.</p></li>
<li><p>The <strong>Clark's divergence distance</strong> is a related metric that can be seen as a squared version of the Canberra distance, averaged and then square-rooted, making it a type of root-mean-square Canberra distance.</p></li>
</ul></div>

<div class="section">
<h3 id="distances-based-on-data-transformation">Distances Based on Data Transformation<a class="anchor" aria-label="anchor" href="#distances-based-on-data-transformation"></a></h3>


<p>A powerful approach to overcoming the problems of the Euclidean distance is
to transform the data before calculating the distance.</p><ul><li><p>The <strong>Hellinger distance</strong> is the Euclidean distance calculated on data that has first been transformed to relative abundance and then square-root transformed. This transformation has the effect of down-weighting the influence of highly abundant species, making the Hellinger distance a robust quantitative metric that is less dominated by outliers than Bray-Curtis.</p></li>
<li><p>The <strong>Chord distance</strong> is the Euclidean distance calculated on data that has been normalized to unit length (i.e., each sample vector is projected onto the surface of a hypersphere). The <strong>Squared Chord distance</strong> is simply the square of this value and ranges from 0 to 2. Hellinger distance is equivalent to the Chord distance of square-root transformed data.</p></li>
<li><p>The <strong>Bhattacharyya</strong> and <strong>Matusita</strong> distances are related metrics that quantify the overlap between two probability distributions. When applied to species abundance vectors (treated as discrete probability distributions), they provide a robust measure of dissimilarity that, like Hellinger, is less sensitive to extreme abundance values.</p></li>
</ul></div>

<div class="section">
<h3 id="specialized-distances">Specialized Distances<a class="anchor" aria-label="anchor" href="#specialized-distances"></a></h3>

<ul><li><p>The <strong>Gower distance</strong> is a flexible metric that averages the difference over all variables, with each term normalized by the range of that variable. This approach makes the Gower distance similar to the Manhattan distance but with range normalization included. It is a useful tool for creating a single distance matrix from a table of species abundance data.</p></li>
<li><p>The <strong>Lorentzian distance</strong>, derived from mathematical physics, is another specialized metric that can be used to alter the contribution of different features (species) to the overall distance, but its application in mainstream ecology is limited.</p></li>
</ul></div>

    </div>
    <div class="section level2">
    <h2 id="calculation">Calculation<a class="anchor" aria-label="anchor" href="#calculation"></a></h2>



<p>Given:</p><ul><li><p>\(P_i\): The proportional abundance of feature \(i\) in sample \(P\).</p></li>
<li><p>\(Q_i\): The proportional abundance of feature \(i\) in sample \(Q\).</p></li>
<li><p>\(n\): The number of features.</p></li>
<li><p>\(p\): For Minkowski only, the root/exponent to apply.</p></li>
<li><p>\(R_i\): For Gower only, the range of the \(i\)-th feature across all samples (max - min).</p></li>
</ul><table class="table table"><tr><td></td><td></td><td></td></tr><tr><td><strong>Bhattacharyya distance</strong></td><td><code>bhattacharyya()</code></td><td>\(-\ln{\sum_{i=1}^{n}\sqrt{P_i * Q_i}}\)</td></tr><tr><td><strong>Canberra distance</strong></td><td><code>canberra()</code></td><td>\(\displaystyle \sum_{i=1}^{n} \frac{|P_i - Q_i|}{P_i + Q_i}\)</td></tr><tr><td><strong>Chebyshev \(L_{\infty}\) distance</strong></td><td><code>chebyshev()</code></td><td>\(\max(|P_i - Q_i|)\)</td></tr><tr><td><strong>Chord distance</strong></td><td><code>chord()</code></td><td>\(\sum_{i=1}^{n}|\sqrt{P_i} - \sqrt{Q_i}|\)</td></tr><tr><td><strong>Clark's divergence distance</strong></td><td><code>clark()</code></td><td>\(\displaystyle \sqrt{\sum_{i=1}^{n}\left(\frac{P_i - Q_i}{P_i + Q_i}\right)^{2}}\)</td></tr><tr><td><strong>Euclidean \(L_2\) distance</strong></td><td><code>euclidean()</code></td><td>\(\sqrt{\sum_{i=1}^{n} (P_i - Q_i)^2}\)</td></tr><tr><td><strong>Gower distance</strong></td><td><code>gower()</code></td><td>\(\displaystyle \frac{1}{n}\sum_{i=1}^{n}\frac{|P_i - Q_i|}{R_i}\)</td></tr><tr><td><strong>Hellinger distance</strong></td><td><code>hellinger()</code></td><td>\(\sqrt{\sum_{i=1}^{n}(\sqrt{P_i} - \sqrt{Q_i})^{2}}\)</td></tr><tr><td><strong>Lorentzian distance</strong></td><td><code>lorentzian()</code></td><td>\(\sum_{i=1}^{n}\ln{(1 + |P_i - Q_i|)}\)</td></tr><tr><td><strong>Manhattan \(L_1\) distance</strong></td><td><code>manhattan()</code></td><td>\(\sum_{i=1}^{n} |P_i - Q_i|\)</td></tr><tr><td><strong>Matusita distance</strong></td><td><code>matusita()</code></td><td>\(\sqrt{\sum_{i=1}^{n}\left(\sqrt{P_i} - \sqrt{Q_i}\right)^2}\)</td></tr><tr><td><strong>Minkowski distance \(L_p\)</strong></td><td><code>minkowski()</code></td><td>\(\sqrt[p]{\sum_{i=1}^{n} (P_i - Q_i)^p}\)</td></tr><tr><td><strong>Squared Chord distance</strong></td><td><code>squared_chord()</code></td><td>\(\sum_{i=1}^{n}\left(\sqrt{P_i} - \sqrt{Q_i}\right)^2\)</td></tr><tr><td><strong>Squared Euclidean distance</strong></td><td><code>squared_euclidean()</code></td><td>\(\sum_{i=1}^{n} (P_i - Q_i)^2\)</td></tr></table></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Bhattacharyya, A. (1943). On a measure of divergence between two statistical
populations defined by their probability distributions. <em>Bulletin of the
Calcutta Mathematical Society</em>, 35, 99–109.</p>
<p>Cha, S.-H. (2007). Comprehensive survey on distance/similarity measures
between probability density functions. <em>International Journal of Mathematical
Models and Methods in Applied Sciences</em>, 1(4), 300–307.</p>
<p>Clark, P. J., &amp; Evans, F. C. (1954). Distance to nearest neighbor as a
measure of spatial relationships in populations. <em>Ecology</em>, 35(4), 445–453.
<a href="https://doi.org/10.2307/1931034" class="external-link">doi:10.2307/1931034</a></p>
<p>Hellinger, E. (1909). Neue Begründung der Theorie quadratischer Formen von
unendlichvielen Veränderlichen. <em>Journal für die reine und angewandte
Mathematik</em>, 1909(136), 210–271. <a href="https://doi.org/10.1515/crll.1909.136.210" class="external-link">doi:10.1515/crll.1909.136.210</a></p>
<p>Minkowski, H. (1896). <em>Geometrie der Zahlen</em>. Teubner,
Leipzig.</p>
<p>Gower JC (1971). A general coefficient of similarity and some of its
properties. <em>Biometrics</em>, 27(4). <a href="https://doi.org/10.2307/2528823" class="external-link">doi:10.2307/2528823</a></p>
<p>Gower JC, Legendre P (1986). Metric and Euclidean Properties of Dissimilarity
Coefficients. <em>Journal of Classification</em>, 3. <a href="https://doi.org/10.1007/BF01896809" class="external-link">doi:10.1007/BF01896809</a></p>
<p>Legendre P, Caceres M (2013). Beta diversity as the variance of community
data: dissimilarity coefficients and partitioning. <em>Ecology Letters</em>, 16(8).
<a href="https://doi.org/10.1111/ele.12141" class="external-link">doi:10.1111/ele.12141</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="co"># Example counts matrix</span></span></span>
<span class="r-in"><span>    <span class="va">ex_counts</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   Saliva Gums Nose Stool</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Streptococcus        162  793   22     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteroides            2    4    2   611</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Corynebacterium        0    0  498     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Haemophilus          180   87    2     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Propionibacterium      1    1  251     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Staphylococcus         0    1  236     1</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">bhattacharyya</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in validate_normalized(&lt;environment&gt;):</span> could not find function "validate_normalized"</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">canberra</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in validate_normalized(&lt;environment&gt;):</span> could not find function "validate_normalized"</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">gower</span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in validate_normalized(&lt;environment&gt;):</span> could not find function "validate_normalized"</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># Only calculate distances for A vs all.</span></span></span>
<span class="r-in"><span>    <span class="fu">bhattacharyya</span><span class="op">(</span><span class="va">ex_counts</span>, pairs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in validate_normalized(&lt;environment&gt;):</span> could not find function "validate_normalized"</span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

