<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Phylogenetic Distances: The UniFrac Family — unweighted_unifrac • ecodive</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Phylogenetic Distances: The UniFrac Family — unweighted_unifrac"><meta name="description" content="The UniFrac family of metrics represents a major conceptual advance for
microbial ecology, integrating phylogenetic information directly into the
calculation of beta diversity. All UniFrac metrics operate on a phylogenetic
tree that includes all taxa observed across the samples being compared. The
fundamental principle is to measure the total branch length of the tree that
is unique to each community, providing a measure of divergent evolution."><meta property="og:description" content="The UniFrac family of metrics represents a major conceptual advance for
microbial ecology, integrating phylogenetic information directly into the
calculation of beta diversity. All UniFrac metrics operate on a phylogenetic
tree that includes all taxa observed across the samples being compared. The
fundamental principle is to measure the total branch length of the tree that
is unique to each community, providing a measure of divergent evolution."><meta property="og:image" content="https://cmmr.github.io/ecodive/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0.9002</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Phylogenetic Distances: The UniFrac Family</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/HEAD/R/bdiv_phylo.r" class="external-link"><code>R/bdiv_phylo.r</code></a></small>
      <div class="d-none name"><code>bdiv_phylo.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The UniFrac family of metrics represents a major conceptual advance for
microbial ecology, integrating phylogenetic information directly into the
calculation of beta diversity. All UniFrac metrics operate on a phylogenetic
tree that includes all taxa observed across the samples being compared. The
fundamental principle is to measure the total branch length of the tree that
is unique to each community, providing a measure of divergent evolution.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">unweighted_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">weighted_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">normalized_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">generalized_unifrac</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  tree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  pairs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">variance_adjusted_unifrac</span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, pairs <span class="op">=</span> <span class="cn">NULL</span>, cpus <span class="op">=</span> <span class="fu"><a href="n_cpus.html">n_cpus</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>A numeric matrix of count data where each column is a sample,
and each row is a feature. Any object coercible with <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code>
can be given here, as well as <code>phyloseq</code>, <code>rbiom</code>,
<code>SummarizedExperiment</code>, and <code>TreeSummarizedExperiment</code> objects.</p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>A <code>phylo</code>-class object representing the phylogenetic tree for
the OTUs in <code>counts</code>. The OTU identifiers given by <code>colnames(counts)</code>
must be present in <code>tree</code>. Can be omitted if a tree is embedded with
the <code>counts</code> object or as <code>attr(counts, 'tree')</code>.</p></dd>


<dt id="arg-pairs">pairs<a class="anchor" aria-label="anchor" href="#arg-pairs"></a></dt>
<dd><p>Which combinations of samples should distances be
calculated for? The default value (<code>NULL</code>) calculates all-vs-all.
Provide a numeric or logical vector specifying positions in the
distance matrix to calculate. See examples.</p></dd>


<dt id="arg-cpus">cpus<a class="anchor" aria-label="anchor" href="#arg-cpus"></a></dt>
<dd><p>How many parallel processing threads should be used. The
default, <code><a href="n_cpus.html">n_cpus()</a></code>, will use all logical CPU cores.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>How much weight to give to relative abundances; a value
between 0 and 1, inclusive. Setting <code>alpha=1</code> is equivalent to
<code>normalized_unifrac()</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>dist</code> object.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<div class="section">
<h3 id="unweighted-and-weighted-unifrac">Unweighted and Weighted UniFrac<a class="anchor" aria-label="anchor" href="#unweighted-and-weighted-unifrac"></a></h3>

<ul><li><p>The original <strong>Unweighted UniFrac</strong> is a qualitative metric. It calculates the fraction of the total branch length in the tree that leads to descendants from only one of the two communities, but not both. Because it operates on presence/absence, it is highly sensitive to changes in rare lineages. The appearance or disappearance of a taxon that represents a deep, unique branch on the tree will result in a large UniFrac distance.</p></li>
<li><p><strong>Weighted UniFrac</strong> is the quantitative counterpart. It weights the branch lengths in the tree by the difference in the proportion of sequences that descend from that branch in each community. This makes it much more sensitive to changes in the abundance of dominant lineages. A small change in a very abundant clade will have a larger impact on the distance than the complete loss of a rare clade.</p></li>
<li><p>The <strong>Normalized Weighted UniFrac</strong> is a simple modification that scales the distance to fall between 0 and 1, facilitating comparisons.</p></li>
</ul></div>

<div class="section">
<h3 id="advanced-unifrac-variants-generalized-and-variance-adjusted">Advanced UniFrac Variants: Generalized and Variance-Adjusted<a class="anchor" aria-label="anchor" href="#advanced-unifrac-variants-generalized-and-variance-adjusted"></a></h3>


<p>The stark difference in sensitivity between unweighted (rare lineages) and
weighted (abundant lineages) UniFrac led to the development of more advanced
versions to capture a wider range of biological changes.</p><ul><li><p><strong>Generalized UniFrac (GUniFrac)</strong> was developed to bridge this gap. It introduces a parameter, \(\alpha\), that tunes the weight given to abundant lineages. When \(\alpha\)=0, GUniFrac is similar to Unweighted UniFrac; when \(\alpha\)=1, it is similar to Weighted UniFrac. By using an intermediate value, such as the recommended \(\alpha\)=0.5, GUniFrac can effectively detect changes in moderately abundant lineages that might be missed by the other two methods, making it a more robust and powerful tool overall.</p></li>
<li><p><strong>Variance-Adjusted Weighted UniFrac (VAW-UniFrac)</strong> addresses a statistical limitation of the original Weighted UniFrac. It recognizes that the observed abundances are estimates subject to sampling variation. VAW-UniFrac adjusts the weight on each branch by its variance, which can increase the statistical power to detect genuine differences between communities, especially when sample sizes are uneven.</p></li>
</ul></div>

    </div>
    <div class="section level2">
    <h2 id="calculation">Calculation<a class="anchor" aria-label="anchor" href="#calculation"></a></h2>



<p>Given \(n\) branches with lengths \(L\) and a pair of samples'
abundances (\(A\) and \(B\)) on each of those branches:</p>
<p><strong>Unweighted UniFrac</strong> <code>unweighted_unifrac()</code></p>
<p>$$D_{unweighted} = \displaystyle \frac{\sum_{i} L_i(|A_i - B_i|)}{\sum_{i} L_i(max(A_i,B_i))}$$</p>
<p>In \(D_{unweighted}\), abundances \(A\) and \(B\) are coded as <code>1</code> or
<code>0</code> to indicate their presence or absence, respectively, on each branch.</p>
<p><strong>Weighted UniFrac</strong> <code>weighted_unifrac()</code></p>
<p>$$D_{weighted} = \displaystyle \sum_{i} L_i|\frac{A_i}{A_T} - \frac{B_i}{B_T}|$$</p>
<p><strong>Normalized Weighted UniFrac</strong> <code>normalized_unifrac()</code></p>
<p>$$D_{normalized} = \displaystyle \frac{\sum_{i} L_i|\frac{A_i}{A_T} - \frac{B_i}{B_T}|}{\sum_{i} L_i(\frac{A_i}{A_T} + \frac{B_i}{B_T})}$$</p>
<p><strong>Generalized UniFrac</strong> <code>generalized_unifrac()</code></p>
<p>$$D_{generalized} = \displaystyle \frac{\sum_{i} L_i(\frac{A_i}{A_T} + \frac{B_i}{B_T})^{\alpha}|\displaystyle \frac{\frac{A_i}{A_T} - \frac{B_i}{B_T}}{\frac{A_i}{A_T} + \frac{B_i}{B_T}} |}{\sum_{i} L_i(\frac{A_i}{A_T} + \frac{B_i}{B_T})^{\alpha}}$$</p>
<p><strong>Variance-Adjusted Weighted UniFrac</strong> <code>variance_adjusted_unifrac()</code></p>
<p>$$D_{varAdjusted} = \displaystyle \frac{\sum_{i} L_i\displaystyle \frac{|\frac{A_i}{A_T} - \frac{B_i}{B_T}|}{\sqrt{(A_i + B_i)(A_T + B_T - A_i - B_i)}} }{\sum_{i} L_i\displaystyle \frac{\frac{A_i}{A_T} + \frac{B_i}{B_T}}{\sqrt{(A_i + B_i)(A_T + B_T - A_i - B_i)}} }$$</p>
<p>See <code><a href="../articles/unifrac.html">vignette('unifrac')</a></code> for details and a worked example.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Lozupone C, Knight R (2005). UniFrac: A new phylogenetic method for comparing
microbial communities.  <em>Applied and Environmental Microbiology</em>, 71(12).
<a href="https://doi.org/10.1128/AEM.71.12.8228-8235.2005" class="external-link">doi:10.1128/AEM.71.12.8228-8235.2005</a></p>
<p>Lozupone CA, Hamady M, Kelley ST, Knight R (2007). Quantitative and
Qualitative \(\beta\) Diversity Measures Lead to Different Insights into
Factors That Structure Microbial Communities.
<em>Applied and Environmental Microbiology</em>, 73(5). <a href="https://doi.org/10.1128/AEM.01996-06" class="external-link">doi:10.1128/AEM.01996-06</a></p>
<p>Chen J, Bittinger K, Charlson ES, Hoffmann C, Lewis J, Wu GD, Collman RG,
Bushman FD, Li H (2012). Associating microbiome composition with
environmental covariates using generalized UniFrac distances.
<em>Bioinformatics</em>, 28(16). <a href="https://doi.org/10.1093/bioinformatics/bts342" class="external-link">doi:10.1093/bioinformatics/bts342</a></p>
<p>Chang Q, Luan Y, Sun F (2011). Variance adjusted weighted UniFrac: a powerful
beta diversity measure for comparing communities based on phylogeny.
<em>BMC Bioinformatics</em>, 12. <a href="https://doi.org/10.1186/1471-2105-12-118" class="external-link">doi:10.1186/1471-2105-12-118</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="co"># Example counts matrix</span></span></span>
<span class="r-in"><span>    <span class="va">ex_counts</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   Saliva Gums Nose Stool</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Streptococcus        162  793   22     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteroides            2    4    2   611</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Corynebacterium        0    0  498     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Haemophilus          180   87    2     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Propionibacterium      1    1  251     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Staphylococcus         0    1  236     1</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">unweighted_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           Saliva       Gums       Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.05759162                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.16279070 0.11162791           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.22325581 0.17209302 0.06046512</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">weighted_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums   37.08021                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose   67.00360  55.56710          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 110.25564 109.96250 110.14056</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">normalized_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4343436                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.7891160 0.6714592          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9707293 0.9868636 0.9925679</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">generalized_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4471644                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.8215129 0.7607876          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9727827 0.9784242 0.9730332</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">variance_adjusted_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Saliva      Gums      Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.4242631                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.7753369 0.5565010          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.9655749 0.9807634 0.9785147</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># Only calculate distances for A vs all.</span></span></span>
<span class="r-in"><span>    <span class="fu">unweighted_unifrac</span><span class="op">(</span><span class="va">ex_counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span>, pairs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           Saliva       Gums       Nose</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gums  0.05759162                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Nose  0.16279070         NA           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Stool 0.22325581         NA         NA</span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

