<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to ecodive • ecodive</title>
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction to ecodive">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.90</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to ecodive</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/main/vignettes/ecodive.Rmd" class="external-link"><code>vignettes/ecodive.Rmd</code></a></small>
      <div class="d-none name"><code>ecodive.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="intoduction">Intoduction<a class="anchor" aria-label="anchor" href="#intoduction"></a>
</h2>
<p>Ecodive calculates ecological diversity metrics. Alpha diversity
metrics provide insight about a single sample’s diversity, whereas beta
diversity metrics indicate how different a pair of samples are from each
other.</p>
<p>In this guide, we’ll use the <code>ex_counts</code> dataset included
with ecodive. <code>ex_counts</code> is a feature table that enumerates
how many times each bacterial genera was observed on different body
sites.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/ecodive/">ecodive</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">ex_counts</span></span>
<span><span class="co">#&gt;                   Saliva Gums Nose Stool</span></span>
<span><span class="co">#&gt; Streptococcus        162  793   22     1</span></span>
<span><span class="co">#&gt; Bacteroides            2    4    2   611</span></span>
<span><span class="co">#&gt; Corynebacterium        0    0  498     1</span></span>
<span><span class="co">#&gt; Haemophilus          180   87    2     1</span></span>
<span><span class="co">#&gt; Propionibacterium      1    1  251     0</span></span>
<span><span class="co">#&gt; Staphylococcus         0    1  236     1</span></span></code></pre></div>
<p>In this example, the ‘features’ in our feature table are genera.
However, your own dataset can use whatever feature makes sense -
species, OTUs, ASVs, or even something completely unrelated to
ecology.</p>
</div>
<div class="section level2">
<h2 id="alpha-diversity">Alpha Diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h2>
<p>Alpha diversity metrics describe how many different genera are
present in a sample. Depending on the metric, this can take into account
the number of unique genera (richness), how evenly the population is
split among genera (evenness), or how distantly related the genera are
(phylogenetic diversity).</p>
<ul>
<li><p>Classic metrics: <code><a href="../reference/chao1.html">chao1()</a></code>, <code><a href="../reference/shannon.html">shannon()</a></code>,
<code><a href="../reference/simpson.html">simpson()</a></code>, <code><a href="../reference/inv_simpson.html">inv_simpson()</a></code></p></li>
<li><p>Phylogenetic metrics: <code><a href="../reference/faith.html">faith()</a></code></p></li>
<li><p>Further reading: <code><a href="../articles/adiv.html">vignette('adiv')</a></code></p></li>
</ul>
</div>
<div class="section level2">
<h2 id="beta-diversity">Beta Diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a>
</h2>
<p>Beta diversity metrics describe how different two samples are, based
on the genera observed in each. Also known as “distance” or
“dissimilarity”. UniFrac metrics incorporate a phylogenetic tree into
this calculation.</p>
<ul>
<li><p>Classic metrics: <code><a href="../reference/bray_curtis.html">bray_curtis()</a></code>,
<code><a href="../reference/canberra.html">canberra()</a></code>, <code><a href="../reference/euclidean.html">euclidean()</a></code>, <code><a href="../reference/gower.html">gower()</a></code>,
<code><a href="../reference/jaccard.html">jaccard()</a></code>, <code><a href="../reference/kulczynski.html">kulczynski()</a></code>,
<code><a href="../reference/manhattan.html">manhattan()</a></code></p></li>
<li><p>Phylogenetic metrics: <code><a href="../reference/unweighted_unifrac.html">unweighted_unifrac()</a></code>,
<code><a href="../reference/weighted_unifrac.html">weighted_unifrac()</a></code>,
<code><a href="../reference/weighted_normalized_unifrac.html">weighted_normalized_unifrac()</a></code>,
<code><a href="../reference/generalized_unifrac.html">generalized_unifrac()</a></code>,
<code><a href="../reference/variance_adjusted_unifrac.html">variance_adjusted_unifrac()</a></code></p></li>
<li><p>Further reading: <code><a href="../articles/bdiv.html">vignette('bdiv')</a></code> and
<code><a href="../articles/unifrac.html">vignette('unifrac')</a></code>.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<div class="section level3">
<h3 id="rarefaction">Rarefaction<a class="anchor" aria-label="anchor" href="#rarefaction"></a>
</h3>
<p>The <code>ex_counts</code> feature table has 345 saliva observations,
but nose has 1011 observations. This unequal sampling depth can cause
systematic biases. Specifically, rare genera will be observed more often
in samples with greater sampling depths, thereby artificially inflating
the observed richness.</p>
<p>The first step then is to rarefy <code>ex_counts</code> so that all
samples have the same number of observations. Rarefying randomly removes
observations from samples with more observations.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; Saliva   Gums   Nose  Stool </span></span>
<span><span class="co">#&gt;    345    886   1011    615 </span></span>
<span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; Saliva   Gums   Nose  Stool </span></span>
<span><span class="co">#&gt;    345    345    345    345 </span></span>
<span></span>
<span><span class="va">counts</span></span>
<span><span class="co">#&gt;                   Saliva Gums Nose Stool</span></span>
<span><span class="co">#&gt; Streptococcus        162  309    6     1</span></span>
<span><span class="co">#&gt; Bacteroides            2    2    0   341</span></span>
<span><span class="co">#&gt; Corynebacterium        0    0  171     1</span></span>
<span><span class="co">#&gt; Haemophilus          180   34    0     1</span></span>
<span><span class="co">#&gt; Propionibacterium      1    0   82     0</span></span>
<span><span class="co">#&gt; Staphylococcus         0    0   86     1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="classic-metrics">Classic Metrics<a class="anchor" aria-label="anchor" href="#classic-metrics"></a>
</h3>
<p>These alpha and beta diversity metrics have been around for 50+ years
and don’t require a phylogenetic tree. The beta diversity functions can
take a <code>weighted = FALSE</code> argument to use only
presence/absence information instead of relative abundances.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Alpha Diversity -------------------</span></span>
<span></span>
<span><span class="fu"><a href="../reference/shannon.html">shannon</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Saliva       Gums       Nose      Stool </span></span>
<span><span class="co">#&gt; 0.74119910 0.35692121 1.10615349 0.07927797 </span></span>
<span></span>
<span></span>
<span><span class="co">## Beta Diversity --------------------</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bray_curtis.html">bray_curtis</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Saliva      Gums      Nose</span></span>
<span><span class="co">#&gt; Gums  0.4260870                    </span></span>
<span><span class="co">#&gt; Nose  0.9797101 0.9826087          </span></span>
<span><span class="co">#&gt; Stool 0.9884058 0.9884058 0.9913043</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bray_curtis.html">bray_curtis</a></span><span class="op">(</span><span class="va">counts</span>, weighted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Saliva      Gums      Nose</span></span>
<span><span class="co">#&gt; Gums  0.1428571                    </span></span>
<span><span class="co">#&gt; Nose  0.5000000 0.7142857          </span></span>
<span><span class="co">#&gt; Stool 0.3333333 0.2500000 0.3333333</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="phylogenetic-metrics">Phylogenetic Metrics<a class="anchor" aria-label="anchor" href="#phylogenetic-metrics"></a>
</h3>
<p>A phylogenetic tree enables alpha and beta diversity metrics to take
into account evolutionary relatedness between the observed genera,
generally giving higher diversity values for samples with more distantly
related genera. Faith (for alpha diversity) and UniFrac (for beta
diversity) are examples of phylogenetic metrics.</p>
<p>The <code>ex_tree</code> object included with ecodive provides the
phylogenetic tree for the genera in <code>ex_counts</code>. For your own
datasets, you can use ecodive’s <code><a href="../reference/read_tree.html">read_tree()</a></code> function to
import a phylogenetic tree from a newick formatted string or file.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Alpha Diversity -------------------</span></span>
<span></span>
<span><span class="fu"><a href="../reference/faith.html">faith</a></span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Saliva   Gums   Nose  Stool </span></span>
<span><span class="co">#&gt;    180    155    101    202 </span></span>
<span></span>
<span></span>
<span><span class="co">## Beta Diversity --------------------</span></span>
<span></span>
<span><span class="fu"><a href="../reference/weighted_normalized_unifrac.html">weighted_normalized_unifrac</a></span><span class="op">(</span><span class="va">counts</span>, tree <span class="op">=</span> <span class="va">ex_tree</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Saliva      Gums      Nose</span></span>
<span><span class="co">#&gt; Gums  0.4328662                    </span></span>
<span><span class="co">#&gt; Nose  0.7928701 0.6767840          </span></span>
<span><span class="co">#&gt; Stool 0.9677535 0.9829736 0.9936121</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="distance-matrices">Distance Matrices<a class="anchor" aria-label="anchor" href="#distance-matrices"></a>
</h3>
<p>Beta diversity functions return a <code>dist</code> object. You can
convert this to a standard R matrix with the <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code>
function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bray_curtis.html">bray_curtis</a></span><span class="op">(</span><span class="va">counts</span>, weighted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">dm</span></span>
<span><span class="co">#&gt;          Saliva      Gums      Nose</span></span>
<span><span class="co">#&gt; Gums  0.1428571                    </span></span>
<span><span class="co">#&gt; Nose  0.5000000 0.7142857          </span></span>
<span><span class="co">#&gt; Stool 0.3333333 0.2500000 0.3333333</span></span>
<span></span>
<span><span class="va">mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span></span>
<span><span class="va">mtx</span></span>
<span><span class="co">#&gt;           Saliva      Gums      Nose     Stool</span></span>
<span><span class="co">#&gt; Saliva 0.0000000 0.1428571 0.5000000 0.3333333</span></span>
<span><span class="co">#&gt; Gums   0.1428571 0.0000000 0.7142857 0.2500000</span></span>
<span><span class="co">#&gt; Nose   0.5000000 0.7142857 0.0000000 0.3333333</span></span>
<span><span class="co">#&gt; Stool  0.3333333 0.2500000 0.3333333 0.0000000</span></span>
<span></span>
<span><span class="va">mtx</span><span class="op">[</span><span class="st">'Saliva'</span>, <span class="st">'Nose'</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 0.5</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
