<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to fastbiom • fastbiom</title>
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction to fastbiom">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fastbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/fastbiom.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/fastbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to fastbiom</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/fastbiom/blob/main/vignettes/fastbiom.Rmd" class="external-link"><code>vignettes/fastbiom.Rmd</code></a></small>
      <div class="d-none name"><code>fastbiom.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="intoduction">Intoduction<a class="anchor" aria-label="anchor" href="#intoduction"></a>
</h2>
<p><code>fastbiom</code> provides fast algorithms for calculating biome
diversity. Most of these algorithms fall into two categories - alpha
diversity and beta diversity.</p>
</div>
<div class="section level2">
<h2 id="alpha-diversity">Alpha Diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h2>
<p>A measure of how many different species are present in a biome.
Depending on the metric, this can take into account the number of unique
species (richness), how evenly the population is split among species
(evenness), or how similar the species are (phylogenetic diversity).</p>
<ul>
<li><p>Classic metrics: <code><a href="../reference/chao1.html">chao1()</a></code>, <code><a href="../reference/shannon.html">shannon()</a></code>,
<code><a href="../reference/simpson.html">simpson()</a></code>, <code><a href="../reference/inv_simpson.html">inv_simpson()</a></code></p></li>
<li><p>Phylogenetic metrics: <code><a href="../reference/faith.html">faith()</a></code></p></li>
<li><p>Further reading: <code><a href="../articles/adiv.html">vignette('adiv')</a></code></p></li>
</ul>
</div>
<div class="section level2">
<h2 id="beta-diversity">Beta Diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a>
</h2>
<p>A measure of how different two biomes are, based on the species
observed. Also known as “distance” or “dissimilarity”. UniFrac metrics
incorporate a phylogenetic tree into this calculation.</p>
<ul>
<li><p>Classic metrics: <code><a href="../reference/bray_curtis.html">bray_curtis()</a></code>,
<code><a href="../reference/canberra.html">canberra()</a></code>, <code><a href="../reference/euclidean.html">euclidean()</a></code>, <code><a href="../reference/gower.html">gower()</a></code>,
<code><a href="../reference/jaccard.html">jaccard()</a></code>, <code><a href="../reference/kulczynski.html">kulczynski()</a></code>,
<code><a href="../reference/manhattan.html">manhattan()</a></code></p></li>
<li><p>Phylogenetic metrics: <code><a href="../reference/unweighted_unifrac.html">unweighted_unifrac()</a></code>,
<code><a href="../reference/weighted_unifrac.html">weighted_unifrac()</a></code>, <code><a href="../reference/normalized_unifrac.html">normalized_unifrac()</a></code>,
<code><a href="../reference/generalized_unifrac.html">generalized_unifrac()</a></code>,
<code><a href="../reference/variance_adjusted_unifrac.html">variance_adjusted_unifrac()</a></code></p></li>
<li><p>Further reading: <code><a href="../articles/bdiv.html">vignette('bdiv')</a></code> and
<code><a href="../articles/unifrac.html">vignette('unifrac')</a></code>.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<div class="section level3">
<h3 id="input-matrix">Input Matrix<a class="anchor" aria-label="anchor" href="#input-matrix"></a>
</h3>
<p>A counts matrix is required for all alpha and beta diversity metrics.
Here we’ll use a simple R matrix, but <code>rbiom</code>,
<code>phyloseq</code>, and similar data structures are also
supported.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  data     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">185</span>, <span class="fl">0</span>, <span class="fl">206</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">900</span>, <span class="fl">0</span>, <span class="fl">100</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">25</span>, <span class="fl">556</span>, <span class="fl">2</span>, <span class="fl">288</span>, <span class="fl">270</span><span class="op">)</span>,</span>
<span>  nrow     <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Streptococcus'</span>, <span class="st">'Corynebacterium'</span>, <span class="st">'Haemophilus'</span>, <span class="st">'Propionibacterium'</span>, <span class="st">'Staphylococcus'</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Saliva'</span>, <span class="st">'Gums'</span>, <span class="st">'Nose'</span><span class="op">)</span> <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts</span></span>
<span><span class="co">#&gt;                   Saliva Gums Nose</span></span>
<span><span class="co">#&gt; Streptococcus        185  900   25</span></span>
<span><span class="co">#&gt; Corynebacterium        0    0  556</span></span>
<span><span class="co">#&gt; Haemophilus          206  100    2</span></span>
<span><span class="co">#&gt; Propionibacterium      1    1  288</span></span>
<span><span class="co">#&gt; Staphylococcus         0    1  270</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; Saliva   Gums   Nose </span></span>
<span><span class="co">#&gt;    392   1002   1141 </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rarefaction">Rarefaction<a class="anchor" aria-label="anchor" href="#rarefaction"></a>
</h3>
<p>The raw counts matrix has 392 saliva observations, but gums and nose
have 1000+ observations. This unequal sampling depth can cause
systematic biases. Specifically, rare species will be observed more
often in biomes with greater sampling depths, thereby artificially
inflating the observed richness.</p>
<p>The first step then is to rarefy the counts matrix so that all biomes
have the same number of observations. This randomly removes observations
from biomes with more observations.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/fastbiom/">fastbiom</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts</span></span>
<span><span class="co">#&gt;                   Saliva Gums Nose</span></span>
<span><span class="co">#&gt; Streptococcus        185  354    9</span></span>
<span><span class="co">#&gt; Corynebacterium        0    0  191</span></span>
<span><span class="co">#&gt; Haemophilus          206   37    2</span></span>
<span><span class="co">#&gt; Propionibacterium      1    1   99</span></span>
<span><span class="co">#&gt; Staphylococcus         0    0   91</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; Saliva   Gums   Nose </span></span>
<span><span class="co">#&gt;    392    392    392</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="classic-metrics">Classic Metrics<a class="anchor" aria-label="anchor" href="#classic-metrics"></a>
</h3>
<p>These metrics have been around for 50+ years and don’t require a
phylogenetic tree. The beta diversity functions can take a
<code>weighted = FALSE</code> argument to use only presence/absence
information instead of relative abundances.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/shannon.html">shannon</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Saliva      Gums      Nose </span></span>
<span><span class="co">#&gt; 0.7077202 0.3301009 1.1504694</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bray_curtis.html">bray_curtis</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt;         Saliva      Gums</span></span>
<span><span class="co">#&gt; Gums 0.4311224          </span></span>
<span><span class="co">#&gt; Nose 0.9693878 0.9693878</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bray_curtis.html">bray_curtis</a></span><span class="op">(</span><span class="va">counts</span>, weighted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Saliva Gums</span></span>
<span><span class="co">#&gt; Gums   0.00     </span></span>
<span><span class="co">#&gt; Nose   0.25 0.25</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="../reference/bray_curtis.html">bray_curtis</a></span><span class="op">(</span><span class="va">counts</span>, weighted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Saliva Gums Nose</span></span>
<span><span class="co">#&gt; Saliva   0.00 0.00 0.25</span></span>
<span><span class="co">#&gt; Gums     0.00 0.00 0.25</span></span>
<span><span class="co">#&gt; Nose     0.25 0.25 0.00</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="phylogenetic-metrics">Phylogenetic Metrics<a class="anchor" aria-label="anchor" href="#phylogenetic-metrics"></a>
</h3>
<p><code><a href="../reference/faith.html">faith()</a></code> and the <code>*_unifrac()</code> functions
require a phylogenetic tree for the species in <code>counts</code>. You
can use <code>fastbiom</code>’s <code><a href="../reference/read_tree.html">read_tree()</a></code> function to
import a phylogenetic tree from a newick formatted string or file.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_tree.html">read_tree</a></span><span class="op">(</span><span class="st">"</span></span>
<span><span class="st">  (Haemophilus:46,</span></span>
<span><span class="st">    ((Streptococcus:18,Staphylococcus:11):12,</span></span>
<span><span class="st">    (Corynebacterium:24,Propionibacterium:13):12):11);"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">#   +----------------46--------------- Haemophilus</span></span>
<span><span class="co">#   |                   </span></span>
<span><span class="co">#   |           +--------18------ Streptococcus</span></span>
<span><span class="co">#   |    +--12--|       </span></span>
<span><span class="co">#   |    |      +--11-- Staphylococcus</span></span>
<span><span class="co">#   +-11-|              </span></span>
<span><span class="co">#        |      +------------24---------- Corynebacterium</span></span>
<span><span class="co">#        +--12--|</span></span>
<span><span class="co">#               +----13--- Propionibacterium</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/faith.html">faith</a></span><span class="op">(</span><span class="va">counts</span>, <span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Saliva   Gums   Nose </span></span>
<span><span class="co">#&gt;    112    112    147 </span></span>
<span></span>
<span><span class="fu"><a href="../reference/normalized_unifrac.html">normalized_unifrac</a></span><span class="op">(</span><span class="va">counts</span>, <span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt;         Saliva      Gums</span></span>
<span><span class="co">#&gt; Gums 0.4408828          </span></span>
<span><span class="co">#&gt; Nose 0.7876058 0.6671201</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
