<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Performance Guide • ecodive</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Performance Guide">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/performance.html">Performance Guide</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Performance Guide</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/main/vignettes/performance.Rmd" class="external-link"><code>vignettes/performance.Rmd</code></a></small>
      <div class="d-none name"><code>performance.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="optimizing-ecodive-for-large-datasets">Optimizing <code>ecodive</code> for Large Datasets<a class="anchor" aria-label="anchor" href="#optimizing-ecodive-for-large-datasets"></a>
</h2>
<p><code>ecodive</code> is engineered for high performance, using
parallelized C code to deliver results quickly and with minimal memory
usage. For most use cases, you can pass your data in any standard R
format (like a <code>matrix</code> or <code>data.frame</code>) and get
fast results.</p>
<p>However, when working with very large datasets—such as those with
thousands of samples or hundreds of thousands of features—you can
achieve noticeably better performance by paying attention to the format
of your input data. These optimizations minimize the internal data
conversion steps that <code>ecodive</code> has to perform.</p>
<div class="section level3">
<h3 id="the-optimal-input-compressed-sparse-matrices">The Optimal Input: Compressed Sparse Matrices<a class="anchor" aria-label="anchor" href="#the-optimal-input-compressed-sparse-matrices"></a>
</h3>
<p><code>ecodive</code> is highly efficient at reformatting data
internally for one-off calculations. You do not need to manually format
your data for single function calls.</p>
<p>However, if you plan to run <strong>multiple <code>ecodive</code>
functions</strong> on the same large dataset, you will see a performance
benefit by converting your data into the optimal format first. This
format is a <strong>column-compressed sparse matrix
(<code>dgCMatrix</code>)</strong> from the <code>Matrix</code> package,
with <strong>samples arranged in columns</strong>.</p>
<p>By doing the conversion once, you prevent <code>ecodive</code> from
having to reformat the data for each subsequent function call.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/ecodive/">ecodive</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assume 'my_counts' is a standard matrix with samples in rows</span></span>
<span><span class="va">my_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">ex_counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert once to the optimal format</span></span>
<span><span class="va">optimal_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">my_counts</span><span class="op">)</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now, run multiple analyses on the pre-formatted object.</span></span>
<span><span class="co"># Remember to specify margin = 2L since samples are in columns.</span></span>
<span><span class="va">shannon_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_functions.html">shannon</a></span><span class="op">(</span><span class="va">optimal_counts</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">simpson_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_functions.html">simpson</a></span><span class="op">(</span><span class="va">optimal_counts</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">observed_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_functions.html">observed</a></span><span class="op">(</span><span class="va">optimal_counts</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="why-sparse-matrices">Why Sparse Matrices?<a class="anchor" aria-label="anchor" href="#why-sparse-matrices"></a>
</h4>
<p>Ecological datasets, particularly from amplicon sequencing (16S, ITS)
or shotgun metagenomics, are typically “sparse.” This means that for any
given sample, the vast majority of species in the full dataset have a
count of zero.</p>
<p>A standard R <code>matrix</code> stores every single value, including
all the zeros. For a large dataset, this can consume a massive amount of
memory. A sparse matrix, on the other hand, only stores the non-zero
values and their locations. This dramatically reduces the memory
footprint.</p>
<p>By providing a <code>dgCMatrix</code> with samples in columns, you
are feeding <code>ecodive</code> data in its preferred native format,
allowing it to skip any internal conversion and proceed directly to
calculations.</p>
</div>
<div class="section level4">
<h4 id="rbiom-users">
<code>rbiom</code> Users<a class="anchor" aria-label="anchor" href="#rbiom-users"></a>
</h4>
<p>If you use the <code>rbiom</code> package to manage your data, you’re
already taking advantage of this! <code>rbiom</code> objects store their
count tables internally as a <code>dgCMatrix</code> with samples in
columns. When you pass an <code>rbiom</code> object to an
<code>ecodive</code> function, it is passed in this optimal format
automatically.</p>
</div>
</div>
<div class="section level3">
<h3 id="handling-data-transformations">Handling Data Transformations<a class="anchor" aria-label="anchor" href="#handling-data-transformations"></a>
</h3>
<p>Many diversity metrics operate on transformed data, most commonly
relative abundances (<code>norm = 'percent'</code>). When you specify a
transformation like <code>norm = 'percent'</code>, <code>ecodive</code>
performs this conversion internally. This adds a small amount of
computational overhead.</p>
<p>If you plan to run many different analyses on the same transformed
data, it is more efficient to perform the transformation once yourself
and then pass the pre-transformed data to <code>ecodive</code>,
specifying <code>norm = 'none'</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inefficient: Transforming the data twice</span></span>
<span><span class="va">shannon_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_functions.html">shannon</a></span><span class="op">(</span><span class="va">optimal_counts</span>, norm <span class="op">=</span> <span class="st">'percent'</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">simpson_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_functions.html">simpson</a></span><span class="op">(</span><span class="va">optimal_counts</span>, norm <span class="op">=</span> <span class="st">'percent'</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># More Efficient: Transform once</span></span>
<span><span class="va">rel_abund</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">optimal_counts</span><span class="op">)</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rel_abund</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">rel_abund</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">shannon_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_functions.html">shannon</a></span><span class="op">(</span><span class="va">rel_abund</span>, norm <span class="op">=</span> <span class="st">'none'</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">simpson_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_functions.html">simpson</a></span><span class="op">(</span><span class="va">rel_abund</span>, norm <span class="op">=</span> <span class="st">'none'</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="the-exception-centered-log-ratio-clr">The Exception: Centered Log-Ratio (CLR)<a class="anchor" aria-label="anchor" href="#the-exception-centered-log-ratio-clr"></a>
</h4>
<p>The one exception to this rule is the centered log-ratio
transformation (<code>norm = 'clr'</code>), used for calculating
Aitchison distance. A standard CLR transformation turns a sparse matrix
into a “dense” matrix (where all zeros become non-zero values), which
dramatically increases memory consumption.</p>
<p><code>ecodive</code>’s internal CLR implementation uses special
techniques to calculate the result while preserving the sparse matrix
structure as much as possible. Therefore, for CLR-based metrics, you
will achieve better performance and lower memory usage by letting
<code>ecodive</code> handle the transformation.</p>
<p><strong>Do this:</strong></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Best for CLR: Let ecodive handle the transformation</span></span>
<span><span class="va">aitchison_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdiv_functions.html">aitchison</a></span><span class="op">(</span><span class="va">optimal_counts</span>, margin <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span></code></pre></div>
<p><strong>Not this:</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inefficient for CLR: Pre-transforming creates a dense matrix</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stat.boogaart.de/compositions/" class="external-link">compositions</a></span><span class="op">)</span></span>
<span><span class="va">dense_matrix</span> <span class="op">&lt;-</span> <span class="fu">clr</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">optimal_counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># Becomes dense</span></span>
<span><span class="va">aitchison_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdiv_functions.html">euclidean</a></span><span class="op">(</span><span class="va">dense_matrix</span>, norm <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h3>
<p>For the best performance with <strong>very large
datasets</strong>:</p>
<ol style="list-style-type: decimal">
<li>Store your count data as a <code>dgCMatrix</code> with samples in
columns.</li>
<li>Use <code>margin = 2L</code> in <code>ecodive</code> function
calls.</li>
<li>For repeated calculations, pre-transform your data (e.g., to
relative abundance) and use <code>norm = 'none'</code>.</li>
<li>
<strong>Exception</strong>: Always let <code>ecodive</code> handle
CLR transformations by using <code>norm='clr'</code>.</li>
</ol>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
