<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>UniFrac Calculations • ecodive</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="UniFrac Calculations">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ecodive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/ecodive.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/adiv.html">Alpha Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv_guide.html">Selecting a Beta Diversity Metric</a></li>
    <li><a class="dropdown-item" href="../articles/bdiv.html">Beta Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/performance.html">Performance Guide</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Calculations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/ecodive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>UniFrac Calculations</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/ecodive/blob/v2.2.1/vignettes/articles/unifrac.Rmd" class="external-link"><code>vignettes/articles/unifrac.Rmd</code></a></small>
      <div class="d-none name"><code>unifrac.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The different UniFrac algorithms are listed below, along with
examples for calculating them.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="input-data">Input Data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<ul>
<li>A numeric matrix with two samples and five species.</li>
<li>A phylogenetic tree for those five species.</li>
</ul>
<table class="table"><tr>
<td>
<table class="table">
<tr>
<th>
</th>
<th>
Sample_A
</th>
<th>
Sample_B
</th>
</tr>
<tr>
<td>
Species_1
</td>
<td align="right">
0
</td>
<td align="right">
1
</td>
</tr>
<tr>
<td>
Species_2
</td>
<td align="right">
0
</td>
<td align="right">
4
</td>
</tr>
<tr>
<td>
Species_3
</td>
<td align="right">
9
</td>
<td align="right">
2
</td>
</tr>
<tr>
<td>
Species_4
</td>
<td align="right">
3
</td>
<td align="right">
8
</td>
</tr>
<tr>
<td>
Species_5
</td>
<td align="right">
3
</td>
<td align="right">
0
</td>
</tr>
</table>
</td>
<td>
<p><img src="figures/unifrac-tree.png" class="r-plt" width="100%"></p>
</td>
</tr></table>
<p><br></p>
</div>
<div class="section level2">
<h2 id="definitions">Definitions<a class="anchor" aria-label="anchor" href="#definitions"></a>
</h2>
<p>The branch indices (green circles) are used for ordering the <span class="math inline">L</span>, <span class="math inline">A</span>, and
<span class="math inline">B</span> arrays. Values for <span class="math inline">L</span> are drawn from the input phylogenetic tree.
Values for <span class="math inline">A</span> and <span class="math inline">B</span> are the total number of species
observations descending from that branch; <span class="math inline">A</span> for Sample_A, and <span class="math inline">B</span> for Sample_B.</p>
<p><img src="figures/unifrac-weights.png" class="r-plt" width="75%" style="display: block; margin: auto;"></p>
<p><br></p>
<table class="table">
<tr>
<td>
<span class="math inline">n = 8</span>
</td>
<td>
Number of branches
</td>
</tr>
<tr>
<td>
<span class="math inline">A = \{9, 0, 0, 0, 9, 6, 3, 3\}</span>
</td>
<td>
Branch <strong>weights</strong> for Sample_A.
</td>
</tr>
<tr>
<td>
<span class="math inline">B = \{7, 5, 1, 4, 2, 8, 8, 0\}</span>
</td>
<td>
Branch <strong>weights</strong> for Sample_B.
</td>
</tr>
<tr>
<td>
<span class="math inline">A_T = 15</span>
</td>
<td>
Total observations for Sample_A.
</td>
</tr>
<tr>
<td>
<span class="math inline">B_T = 15</span>
</td>
<td>
Total observations for Sample_B.
</td>
</tr>
<tr>
<td>
<span class="math inline">L = \{0.2, 0.4, 0.8, 0.5, 0.9, 0.6, 0.7,
0.3\}</span>
</td>
<td>
The branch <strong>lengths</strong>.
</td>
</tr>
</table>
<p><br></p>
</div>
<div class="section level2">
<h2 id="unweighted">Unweighted<a class="anchor" aria-label="anchor" href="#unweighted"></a>
</h2>
<ul>
<li>
<a href="https://doi.org/10.1128/AEM.71.12.8228-8235.2005" class="external-link">Lozupone
et al, 2005</a>: Unweighted UniFrac</li>
<li>R Package <a href="https://cran.r-project.org/package=ecodive" class="external-link">ecodive</a>:
<code><a href="../reference/bdiv_functions.html">unweighted_unifrac()</a></code>
</li>
<li>R Package <a href="https://doi.org/10.32614/CRAN.package.abdiv" class="external-link">abdiv</a>:
<code><a href="../reference/bdiv_functions.html">unweighted_unifrac()</a></code>
</li>
<li>R Package <a href="https://doi.org/doi:10.18129/B9.bioc.phyloseq" class="external-link">phyloseq</a>:
<code>UniFrac(weighted=FALSE)</code>
</li>
<li>
<a href="https://qiime2.org/" class="external-link">qiime2</a>
<code>qiime diversity beta-phylogenetic --p-metric unweighted_unifrac</code>
</li>
<li>
<a href="https://mothur.org/" class="external-link">mothur</a>:
<code>unifrac.unweighted()</code>
</li>
</ul>
<p>First, transform A and B into presence (1) and absence (0)
indicators.</p>
<table style="width:75%" class="table"><tr>
<td style="border-width:0">
<p><span class="math display">\begin{align*}
A  &amp;= \{9, 0, 0, 0, 9, 6, 3, 3\} \\
A' &amp;= \{1, 0, 0, 0, 1, 1, 1, 1\}
\end{align*}</span></p>
</td>
<td style="border-width:0">
<p><span class="math display">\begin{align*}
B  &amp;= \{7, 5, 1, 4, 2, 8, 8, 0\} \\
B' &amp;= \{1, 1, 1, 1, 1, 1, 1, 0\}
\end{align*}</span></p>
</td>
</tr></table>
<p>Then apply the formula:</p>
<p><span class="math display">\begin{align*}
U &amp;= \displaystyle \frac{\sum_{i = 1}^{n} L_i(|A'_i -
B'_i|)}{\sum_{i = 1}^{n} L_i(max(A'_i,B'_i))}  \\ \\
U &amp;= \displaystyle \frac{L_1(|A'_1-B'_1|) +
L_2(|A'_2-B'_2|) + \cdots +
L_n(|A'_n-B'_n|)}{L_1(max(A'_1,B'_1)) +
L_2(max(A'_2,B'_2)) + \cdots + L_n(max(A'_n,B'_n))}  \\
\\
U &amp;= \displaystyle \frac{0.2(|1-1|) + 0.4(|0-1|) + \cdots +
0.3(|1-0|)}{0.2(max(1,1)) + 0.4(max(0,1)) + \cdots + 0.3(max(1,0))}  \\
\\
U &amp;= \displaystyle \frac{0.2(0) + 0.4(1) + 0.8(1) + 0.5(1) + 0.9(0)
+ 0.6(0) + 0.7(0) + 0.3(1)}{0.2(1) + 0.4(1) + 0.8(1) + 0.5(1) + 0.9(1) +
0.6(1) + 0.7(1) + 0.3(1)}  \\ \\
U &amp;= \displaystyle \frac{0.4 + 0.8 + 0.5 + 0.3}{0.2 + 0.4 + 0.8 +
0.5 + 0.9 + 0.6 + 0.7 + 0.3}  \\ \\
U &amp;= \displaystyle \frac{2}{4.4}  \\ \\
U &amp;= 0.4545455
\end{align*}</span></p>
</div>
<div class="section level2">
<h2 id="weighted">Weighted<a class="anchor" aria-label="anchor" href="#weighted"></a>
</h2>
<ul>
<li>
<a href="https://doi.org/10.1128/AEM.01996-06" class="external-link">Lozupone et al,
2007</a>: Raw Weighted UniFrac</li>
<li>R Package <a href="https://cran.r-project.org/package=ecodive" class="external-link">ecodive</a>:
<code><a href="../reference/bdiv_functions.html">weighted_unifrac()</a></code>
</li>
<li>R Package <a href="https://doi.org/10.32614/CRAN.package.abdiv" class="external-link">abdiv</a>:
<code><a href="../reference/bdiv_functions.html">weighted_unifrac()</a></code>
</li>
<li>R Package <a href="https://doi.org/doi:10.18129/B9.bioc.phyloseq" class="external-link">phyloseq</a>:
<code>UniFrac(weighted=TRUE, normalized=FALSE)</code>
</li>
<li>
<a href="https://qiime2.org/" class="external-link">qiime2</a>
<code>qiime diversity beta-phylogenetic --p-metric weighted_unifrac</code>
</li>
</ul>
<p><span class="math display">\begin{align*}
W &amp;= \sum_{i = 1}^{n} L_i|\frac{A_i}{A_T} - \frac{B_i}{B_T}|  \\ \\
W &amp;= L_1|\frac{A_1}{A_T} - \frac{B_1}{B_T}| + L_2|\frac{A_2}{A_T} -
\frac{B_2}{B_T}| + \cdots + L_n|\frac{A_n}{A_T} - \frac{B_n}{B_T}|  \\
\\
W &amp;= 0.2|\frac{9}{15} - \frac{7}{15}| + 0.4|\frac{0}{15} -
\frac{5}{15}| + \cdots + 0.3|\frac{3}{15} - \frac{0}{15}|  \\ \\
W &amp;= 0.02\overline{6} + 0.1\overline{3} + 0.05\overline{3} +
0.1\overline{3} + 0.42 + 0.08 + 0.2\overline{3} + 0.06  \\ \\
W &amp;= 1.14
\end{align*}</span></p>
</div>
<div class="section level2">
<h2 id="normalized">Normalized<a class="anchor" aria-label="anchor" href="#normalized"></a>
</h2>
<ul>
<li>
<a href="https://doi.org/10.1128/AEM.01996-06" class="external-link">Lozupone et al,
2007</a>: Normalized Weighted UniFrac</li>
<li>R Package <a href="https://cran.r-project.org/package=ecodive" class="external-link">ecodive</a>:
<code><a href="../reference/bdiv_functions.html">normalized_unifrac()</a></code>
</li>
<li>R Package <a href="https://doi.org/10.32614/CRAN.package.abdiv" class="external-link">abdiv</a>:
<code>weighted_normalized_unifrac()</code>
</li>
<li>R Package <a href="https://doi.org/doi:10.18129/B9.bioc.phyloseq" class="external-link">phyloseq</a>:
<code>UniFrac(weighted=TRUE, normalized=TRUE)</code>
</li>
<li>
<a href="https://qiime2.org/" class="external-link">qiime2</a>
<code>qiime diversity beta-phylogenetic --p-metric weighted_normalized_unifrac</code>
</li>
<li>
<a href="https://mothur.org/" class="external-link">mothur</a>:
<code>unifrac.weighted()</code>
</li>
</ul>
<p><span class="math display">\begin{align*}

N &amp;= \displaystyle \frac
  {\sum_{i = 1}^{n} L_i|\frac{A_i}{A_T} - \frac{B_i}{B_T}|}
  {\sum_{i = 1}^{n} L_i(\frac{A_i}{A_T} + \frac{B_i}{B_T})}
  \\ \\
  
N &amp;= \displaystyle \frac
  {L_1|\frac{A_1}{A_T} - \frac{B_1}{B_T}| + L_2|\frac{A_2}{A_T} -
\frac{B_2}{B_T}| + \cdots + L_n|\frac{A_n}{A_T} - \frac{B_n}{B_T}|}
  {L_1(\frac{A_1}{A_T} + \frac{B_1}{B_T}) + L_2(\frac{A_2}{A_T} +
\frac{B_2}{B_T}) + \cdots + L_n(\frac{A_n}{A_T} + \frac{B_n}{B_T})}
  \\ \\
  
N &amp;= \displaystyle \frac
  {0.2|\frac{9}{15} - \frac{7}{15}| + 0.4|\frac{0}{15} - \frac{5}{15}| +
\cdots + 0.3|\frac{3}{15} - \frac{0}{15}|}
  {0.2(\frac{9}{15} + \frac{7}{15}) + 0.4(\frac{0}{15} + \frac{5}{15}) +
\cdots + 0.3(\frac{3}{15} + \frac{0}{15})}
  \\ \\
  
N &amp;= \displaystyle \frac
  {0.02\overline{6} + 0.1\overline{3} + 0.05\overline{3} +
0.1\overline{3} + 0.42 + 0.08 + 0.2\overline{3} + 0.06}
  {0.21\overline{3} + 0.1\overline{3} + 0.05\overline{3} +
0.1\overline{3} + 0.66 + 0.56 + 0.51\overline{3} + 0.06}
  \\ \\
  
N &amp;= \displaystyle \frac{1.14}{2.326667}  \\ \\
N &amp;= 0.4899713
\end{align*}</span></p>
</div>
<div class="section level2">
<h2 id="generalized">Generalized<a class="anchor" aria-label="anchor" href="#generalized"></a>
</h2>
<ul>
<li>
<a href="https://doi.org/10.1093/bioinformatics/bts342" class="external-link">Chen et
al. 2012</a>: Generalized UniFrac</li>
<li>R Package <a href="https://cran.r-project.org/package=ecodive" class="external-link">ecodive</a>:
<code>generalized_unifrac(alpha = 0.5)</code>
</li>
<li>R Package <a href="https://doi.org/10.32614/CRAN.package.abdiv" class="external-link">abdiv</a>:
<code>generalized_unifrac(alpha = 0.5)</code>
</li>
<li>R Package <a href="https://doi.org/10.32614/CRAN.package.GUniFrac" class="external-link">GUniFrac</a>:
<code>GUniFrac(alpha = 0.5)</code>
</li>
<li>
<a href="https://qiime2.org/" class="external-link">qiime2</a>
<code>qiime diversity beta-phylogenetic --p-metric generalized_unifrac -a 0.5</code>
</li>
</ul>
<p><span class="math display">\begin{align*}

G &amp;= \displaystyle \frac
  {\sum_{i = 1}^{n} L_i(\frac{A_i}{A_T} + \frac{B_i}{B_T})^{\alpha}
    |\displaystyle \frac
      {\frac{A_i}{A_T} - \frac{B_i}{B_T}}
      {\frac{A_i}{A_T} + \frac{B_i}{B_T}} |}
  {\sum_{i = 1}^{n} L_i(\frac{A_i}{A_T} + \frac{B_i}{B_T})^{\alpha}}  \\
\\

G &amp;= \displaystyle \frac
  {
    L_1(\frac{A_1}{A_T} + \frac{B_1}{B_T})^{0.5}
      |\displaystyle \frac
        {\frac{A_1}{A_T} - \frac{B_1}{B_T}}
        {\frac{A_1}{A_T} + \frac{B_1}{B_T}}| +
    \cdots +
    L_n(\frac{A_n}{A_T} + \frac{B_n}{B_T})^{0.5}
      |\displaystyle \frac
        {\frac{A_n}{A_T} - \frac{B_n}{B_T}}
        {\frac{A_n}{A_T} + \frac{B_n}{B_T}}|
  }{
    L_1(\frac{A_1}{A_T} + \frac{B_1}{B_T})^{0.5} +
    \cdots +
    L_n(\frac{A_n}{A_T} + \frac{B_n}{B_T})^{0.5}
  }  
  \\ \\

G &amp;= \displaystyle \frac
  {
    0.2(\frac{9}{15} + \frac{7}{15})^{0.5}
      |\displaystyle \frac
        {\frac{9}{15} - \frac{7}{15}}
        {\frac{9}{15} + \frac{7}{15}}| +
    \cdots +
    0.3(\frac{3}{15} + \frac{0}{15})^{0.5}
      |\displaystyle \frac
        {\frac{3}{15} - \frac{0}{15}}
        {\frac{3}{15} + \frac{0}{15}}|
  }{
    0.2(\frac{9}{15} + \frac{7}{15})^{0.5} +
    \cdots +
    0.3(\frac{3}{15} + \frac{0}{15})^{0.5}
  }  
  \\ \\
  
G &amp;\approx \displaystyle \frac
  {0.03 + 0.23 + 0.21 + 0.26 + 0.49 + 0.08 + 0.27 + 0.13}
  {0.21 + 0.23 + 0.21 + 0.26 + 0.77 + 0.58 + 0.60 + 0.13}
  \\ \\
  
G &amp;= \displaystyle \frac{1.701419}{2.986235}  \\ \\
G &amp;= 0.569754

\end{align*}</span></p>
</div>
<div class="section level2">
<h2 id="variance-adjusted">Variance Adjusted<a class="anchor" aria-label="anchor" href="#variance-adjusted"></a>
</h2>
<ul>
<li>
<a href="https://doi.org/10.1186/1471-2105-12-118" class="external-link">Chang et al,
2011</a>: Variance Adjusted Weighted (VAW) UniFrac</li>
<li>R Package <a href="https://cran.r-project.org/package=ecodive" class="external-link">ecodive</a>:
<code><a href="../reference/bdiv_functions.html">variance_adjusted_unifrac()</a></code>
</li>
<li>R Package <a href="https://doi.org/10.32614/CRAN.package.abdiv" class="external-link">abdiv</a>:
<code><a href="../reference/bdiv_functions.html">variance_adjusted_unifrac()</a></code>
</li>
<li>
<a href="https://qiime2.org/" class="external-link">qiime2</a>
<code>qiime diversity beta-phylogenetic --p-metric weighted_normalized_unifrac --p-variance-adjusted</code>
</li>
</ul>
<p><span class="math display">\begin{align*}

V &amp;= \displaystyle \frac
  {\sum_{i = 1}^{n} L_i\displaystyle \frac
    {|\frac{A_i}{A_T} - \frac{B_i}{B_T}|}
    {\sqrt{(A_i + B_i)(A_T + B_T - A_i - B_i)}} }
  {\sum_{i = 1}^{n} L_i\displaystyle \frac
    {\frac{A_i}{A_T} + \frac{B_i}{B_T}}
    {\sqrt{(A_i + B_i)(A_T + B_T - A_i - B_i)}} }  \\ \\

V &amp;= \displaystyle \frac
  {
    L_1\displaystyle \frac
    {|\frac{A_1}{A_T} - \frac{B_1}{B_T}|}
    {\sqrt{(A_1 + B_1)(A_T + B_T - A_1 - B_1)}} +
    \cdots +
    L_n\displaystyle \frac
    {|\frac{A_n}{A_T} - \frac{B_n}{B_T}|}
    {\sqrt{(A_n + B_n)(A_T + B_T - A_n - B_n)}}
  }{
    L_1\displaystyle \frac
    {\frac{A_1}{A_T} + \frac{B_1}{B_T}}
    {\sqrt{(A_1 + B_1)(A_T + B_T - A_1 - B_1)}} +
    \cdots +
    L_n\displaystyle \frac
    {\frac{A_n}{A_T} + \frac{B_n}{B_T}}
    {\sqrt{(A_n + B_n)(A_T + B_T - A_n - B_n)}}
  }  
  \\ \\

V &amp;= \displaystyle \frac
  {
    0.2\displaystyle \frac
    {|\frac{9}{15} - \frac{7}{15}|}
    {\sqrt{(9 + 7)(15 + 15 - 9 - 7)}} +
    \cdots +
    0.3\displaystyle \frac
    {|\frac{3}{15} - \frac{0}{15}|}
    {\sqrt{(3 + 0)(15 + 15 - 3 - 0)}}
  }{
    0.2\displaystyle \frac
    {\frac{9}{15} + \frac{7}{15}}
    {\sqrt{(9 + 7)(15 + 15 - 9 - 7)}} +
    \cdots +
    0.3\displaystyle \frac
    {\frac{3}{15} + \frac{0}{15}}
    {\sqrt{(3 + 0)(15 + 15 - 3 - 0)}}
  }  
  \\ \\
  
V &amp;\approx \displaystyle \frac
  {0.002 + 0.012 + 0.010 + 0.013 + 0.029 + 0.005 + 0.016 + 0.007}
  {0.014 + 0.012 + 0.010 + 0.013 + 0.046 + 0.037 + 0.036 + 0.007}
  \\ \\
  
V &amp;= \displaystyle \frac{4.09389}{4.174402}  \\ \\
V &amp;= 0.9807128

\end{align*}</span></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
